{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6306851f-6603-488f-902e-42748dce976e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import shlex\n",
    "import shutil\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3a5baf70-e74c-417f-bf2f-d9ae35bbe3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(\"data/gwas\", ignore_errors=True)\n",
    "gwas_root = pathlib.Path(\"data/gwas/\")\n",
    "gwas_root.mkdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0a32e913-83bb-4e55-a0c6-fd9f38bc1569",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['icd', 'phecodes', 'boolean', 'approx_phecodes', 'approx_anon_phecodes_0005', 'approx_anon_phecodes_0010', 'approx_anon_phecodes_0015', 'approx_anon_phecodes_0020', 'approx_anon_phecodes_0050', 'approx_anon_phecodes_0100', 'approx_boolean', 'approx_anon_boolean_0005', 'approx_anon_boolean_0010', 'approx_anon_boolean_0015', 'approx_anon_boolean_0020', 'approx_anon_boolean_0050', 'approx_anon_boolean_0100']\n"
     ]
    }
   ],
   "source": [
    "pheno_root = pathlib.Path(\"data/pheno\")\n",
    "pheno_paths = [\n",
    "    p for p in pheno_root.glob(\"*.tsv\") \n",
    "    if not p.stem.startswith(\"anon_\")\n",
    "]\n",
    "print([p.stem for p in pheno_paths])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c7b23c0-e60b-4c6f-992f-fc47e3abf232",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GWAS for icd\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/icd/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/icd/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/icd.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:48:39 2024\n",
      "1031943 MiB RAM detected, ~951148 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "340 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-340: done.\n",
      "Results written to data/gwas/icd/result.<phenotype name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:49:12 2024\n",
      "Running GWAS for phecodes\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/phecodes/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/phecodes/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/phecodes.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:49:12 2024\n",
      "1031943 MiB RAM detected, ~951110 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "140 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-140: done.\n",
      "Results written to data/gwas/phecodes/result.<phenotype name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:49:25 2024\n",
      "Running GWAS for boolean\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/boolean/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/boolean/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/boolean.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:49:25 2024\n",
      "1031943 MiB RAM detected, ~951100 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/boolean/result.<phenotype name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:49:53 2024\n",
      "Running GWAS for approx_phecodes\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_phecodes/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_phecodes/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_phecodes.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:49:53 2024\n",
      "1031943 MiB RAM detected, ~951086 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "140 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-140: done.\n",
      "Results written to data/gwas/approx_phecodes/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:50:07 2024\n",
      "Running GWAS for approx_anon_phecodes_0005\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0005/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0005/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0005.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:50:07 2024\n",
      "1031943 MiB RAM detected, ~951087 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0005/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:50:20 2024\n",
      "Running GWAS for approx_anon_phecodes_0010\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0010/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0010/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0010.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:50:20 2024\n",
      "1031943 MiB RAM detected, ~951091 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0010/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:50:33 2024\n",
      "Running GWAS for approx_anon_phecodes_0015\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0015/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0015/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0015.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:50:34 2024\n",
      "1031943 MiB RAM detected, ~951091 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0015/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:50:47 2024\n",
      "Running GWAS for approx_anon_phecodes_0020\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0020/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0020/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0020.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:50:47 2024\n",
      "1031943 MiB RAM detected, ~951089 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0020/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:51:00 2024\n",
      "Running GWAS for approx_anon_phecodes_0050\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0050/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0050/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0050.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:51:00 2024\n",
      "1031943 MiB RAM detected, ~951091 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0050/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:51:13 2024\n",
      "Running GWAS for approx_anon_phecodes_0100\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_phecodes_0100/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_phecodes_0100/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_phecodes_0100.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:51:14 2024\n",
      "1031943 MiB RAM detected, ~951096 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "144 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-144: done.\n",
      "Results written to data/gwas/approx_anon_phecodes_0100/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:51:27 2024\n",
      "Running GWAS for approx_boolean\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_boolean/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_boolean/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_boolean.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:51:27 2024\n",
      "1031943 MiB RAM detected, ~951095 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_boolean/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:51:54 2024\n",
      "Running GWAS for approx_anon_boolean_0005\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0005/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0005/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0005.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:51:54 2024\n",
      "1031943 MiB RAM detected, ~951100 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0005/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:52:21 2024\n",
      "Running GWAS for approx_anon_boolean_0010\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0010/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0010/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0010.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:52:21 2024\n",
      "1031943 MiB RAM detected, ~951097 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0010/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:52:48 2024\n",
      "Running GWAS for approx_anon_boolean_0015\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0015/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0015/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0015.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:52:49 2024\n",
      "1031943 MiB RAM detected, ~951092 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0015/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:53:15 2024\n",
      "Running GWAS for approx_anon_boolean_0020\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0020/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0020/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0020.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:53:16 2024\n",
      "1031943 MiB RAM detected, ~951096 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0020/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:53:44 2024\n",
      "Running GWAS for approx_anon_boolean_0050\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0050/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0050/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0050.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:53:45 2024\n",
      "1031943 MiB RAM detected, ~951088 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0050/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:54:11 2024\n",
      "Running GWAS for approx_anon_boolean_0100\n",
      "PLINK v2.0.0-a.6.0LM AVX2 Intel (11 Nov 2024)      cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/approx_anon_boolean_0100/result.log.\n",
      "Options in effect:\n",
      "  --glm zs hide-covar allow-no-covars\n",
      "  --out data/gwas/approx_anon_boolean_0100/result\n",
      "  --pfile ../../data/geno/ukb_wb_subsampled\n",
      "  --pheno data/pheno/approx_anon_boolean_0100.tsv\n",
      "  --threads 110\n",
      "\n",
      "Start time: Fri Dec  6 12:54:12 2024\n",
      "1031943 MiB RAM detected, ~951089 available; reserving 515971 MiB for main\n",
      "workspace.\n",
      "Using up to 110 threads (change this with --threads).\n",
      "429954 samples (232741 females, 197213 males; 429954 founders) loaded from\n",
      "../../data/geno/ukb_wb_subsampled.psam.\n",
      "10000 variants loaded from ../../data/geno/ukb_wb_subsampled.pvar.\n",
      "300 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-240: done.\n",
      "--glm linear regression on quantitative phenotypes #241-300: done.\n",
      "Results written to data/gwas/approx_anon_boolean_0100/result.<phenotype\n",
      "name>.glm.linear.zst .\n",
      "End time: Fri Dec  6 12:54:38 2024\n"
     ]
    }
   ],
   "source": [
    "for pheno_path in pheno_paths:\n",
    "    pheno_type = pheno_path.stem\n",
    "    print(f\"Running GWAS for {pheno_type}\")\n",
    "\n",
    "    output_dir = gwas_root.joinpath(pheno_type)\n",
    "    output_dir.mkdir(exist_ok=True)\n",
    "\n",
    "    command = f\"\"\"\n",
    "    plink2 \\\n",
    "        --pfile ../../data/geno/ukb_wb_subsampled \\\n",
    "        --pheno {pheno_path.as_posix()} \\\n",
    "        --glm zs hide-covar allow-no-covars \\\n",
    "        --threads 110 \\\n",
    "        --out {output_dir.joinpath('result').as_posix()}\n",
    "    \"\"\"\n",
    "    result = subprocess.run(shlex.split(command))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
