{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5992c318-1e26-46c4-bc5d-7f7dedfe6b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import shlex\n",
    "import subprocess\n",
    "from string import Template\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import tqdm.notebook as tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b5e7d89-07de-4530-ae87-d8fd6f224d0e",
   "metadata": {},
   "source": [
    "# Format simulated data\n",
    "\n",
    "## Recode the chromosome in genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "16a3baa1-8f80-440b-ae73-6e8692390a2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pl.read_csv(\"data/simulated-phenotypes/Genotypes.bim\", separator=\"\\t\", has_header=False)\n",
    "    .with_columns(pl.lit(1).alias(\"column_1\"))\n",
    "    .write_csv(\"data/simulated-phenotypes/Genotypes.bim\", separator=\"\\t\", include_header=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebf3f841-71d6-4eed-a500-6d18d5773f78",
   "metadata": {},
   "source": [
    "## Add names to the phenotype file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab626849-12b0-45a7-a243-9d3f3188f024",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pl.read_csv(\"data/simulated-phenotypes/Ysim_plink.txt\", separator=\"\\t\", has_header=False)\n",
    "    .rename({\n",
    "        \"column_1\": \"FID\",\n",
    "        \"column_2\": \"IID\",\n",
    "        \"column_3\": \"feature_01\",\n",
    "        \"column_4\": \"feature_02\",\n",
    "        \"column_5\": \"feature_03\",\n",
    "        \"column_6\": \"feature_04\",\n",
    "        \"column_7\": \"feature_05\",\n",
    "        \"column_8\": \"feature_06\",\n",
    "        \"column_9\": \"feature_07\",\n",
    "        \"column_10\": \"feature_08\",\n",
    "        \"column_11\": \"feature_09\",\n",
    "        \"column_12\": \"feature_10\",\n",
    "    })\n",
    "    .write_csv(\"data/simulated-phenotypes/Ysim_plink_names.txt\", separator=\"\\t\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "120915df-52ee-4f61-b94c-35b6801b8a63",
   "metadata": {},
   "source": [
    "## Combine phenotypes and covariates into a single file\n",
    "\n",
    "Needed for SAIGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e8e0c601-135f-47b8-8c8f-258d839b26be",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pl.read_csv(\"data/simulated-phenotypes/Ysim_plink_names.txt\", separator=\"\\t\")\n",
    "    .join(\n",
    "        pl.read_csv(\"data/simulated-phenotypes/Covs_plink.txt\", separator=\"\\t\"), \n",
    "        on=[\"FID\", \"IID\"]\n",
    "    )\n",
    "    .write_csv(\"data/simulated-phenotypes/Ysim_saige.txt\", separator=\"\\t\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f60cd8-11de-404f-8ef0-5937fbdcbb1d",
   "metadata": {},
   "source": [
    "# Setup directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "97e8279d-7b44-4acb-9305-ff93a87fba09",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathlib.Path(\"data/grm/\").mkdir(exist_ok=True)\n",
    "\n",
    "pathlib.Path(\"data/projected-phenotypes/\").mkdir(exist_ok=True)\n",
    "\n",
    "gwas_root = pathlib.Path(\"data/gwas/\")\n",
    "gwas_root.mkdir(exist_ok=True)\n",
    "\n",
    "gwas_methods = [\"ols_univariate\", \"ols_multivariate\", \"fastgwa\", \"saige\", \"regenie\"]\n",
    "for gwas_method in gwas_methods:\n",
    "    gwas_root.joinpath(gwas_method).mkdir(exist_ok=True)\n",
    "    gwas_root.joinpath(gwas_method).joinpath(\"features\").mkdir(exist_ok=True)\n",
    "    gwas_root.joinpath(gwas_method).joinpath(\"projections\").mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd778cb4-14a4-43b9-9101-9db2d664c73b",
   "metadata": {},
   "source": [
    "# Random linear combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7d3e5525-2806-41b0-9950-1b2fb3f71676",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "n_features = 10\n",
    "n_projections = 10\n",
    "\n",
    "projection_matrix = np.random.normal(size=(n_features, n_projections))\n",
    "projection_names = [f\"projection_{i:03}\" for i in range(1, n_projections + 1)]\n",
    "feature_names = [f\"feature_{i:02}\" for i in range(1, n_features + 1)]\n",
    "projection_df = pd.DataFrame(projection_matrix, columns=projection_names, index=feature_names)\n",
    "\n",
    "feature_df = (\n",
    "    pl.read_csv(\"data/simulated-phenotypes/Ysim_plink_names.txt\", separator=\"\\t\")\n",
    "    .to_pandas()\n",
    "    .set_index([\"FID\", \"IID\"])\n",
    ")\n",
    "\n",
    "# Normalize the coefficients so that projections have unit variance\n",
    "projection_df /= np.sqrt(np.diag(projection_df.T @ feature_df.cov() @ projection_df))\n",
    "\n",
    "projected_df = feature_df @ projection_df\n",
    "\n",
    "assert np.allclose(feature_df.mean(), 0)\n",
    "assert np.allclose(projected_df.mean(), 0)\n",
    "\n",
    "assert np.allclose(np.diag(feature_df.cov()), 1)\n",
    "assert np.allclose(np.diag(projected_df.cov()), 1)\n",
    "\n",
    "(\n",
    "    projection_df\n",
    "    .reset_index()\n",
    "    .pipe(pl.DataFrame)\n",
    "    .write_csv(\"data/projected-phenotypes/projection_matrix.tsv\", separator=\"\\t\")\n",
    ")\n",
    "\n",
    "(\n",
    "    projected_df\n",
    "    .reset_index()\n",
    "    .pipe(pl.DataFrame)\n",
    "    .write_csv(\"data/projected-phenotypes/phenotypes.tsv\", separator=\"\\t\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ec3bd819-ed1b-4519-baf2-6d1082146a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pl.read_csv(\"data/projected-phenotypes/phenotypes.tsv\", separator=\"\\t\")\n",
    "    .join(\n",
    "        pl.read_csv(\"data/simulated-phenotypes/Covs_plink.txt\", separator=\"\\t\"), \n",
    "        on=[\"FID\", \"IID\"]\n",
    "    )\n",
    "    .write_csv(\"data/projected-phenotypes/phenotypes_saige.tsv\", separator=\"\\t\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9b1dc7-0636-4c38-8894-32f79912fdfb",
   "metadata": {},
   "source": [
    "# Make GRM\n",
    "\n",
    "## GCTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fb87c15-9196-4c57-b0e4-c7941e19a240",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\n",
      "* GCTA (Genome-wide Complex Trait Analysis)\n",
      "* Version v1.94.1 Mac\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang jian.yang@westlake.edu.cn\n",
      "* MIT License\n",
      "*******************************************************************\n",
      "Analysis started at 16:46:09 PST on Wed Nov 13 2024.\n",
      "Hostname: awlpdcg540mz\n",
      "\n",
      "Options: \n",
      " \n",
      "--bfile data/simulated-phenotypes/Genotypes \n",
      "--chr 1 \n",
      "--make-grm \n",
      "--out data/grm/gcta \n",
      "\n",
      "Note: GRM is computed using the SNPs on the autosomes.\n",
      "Reading PLINK FAM file from [data/simulated-phenotypes/Genotypes.fam]...\n",
      "10000 individuals to be included from FAM file.\n",
      "10000 individuals to be included. 0 males, 0 females, 10000 unknown.\n",
      "Reading PLINK BIM file from [data/simulated-phenotypes/Genotypes.bim]...\n",
      "10000 SNPs to be included from BIM file(s).\n",
      "Computing the genetic relationship matrix (GRM) v2 ...\n",
      "Subset 1/1, no. subject 1-10000\n",
      "  10000 samples, 10000 markers, 50005000 GRM elements\n",
      "IDs for the GRM file have been saved in the file [data/grm/gcta.grm.id]\n",
      "Computing GRM...\n",
      "  100% finished in 2.4 sec\n",
      "10000 SNPs have been processed.\n",
      "  Used 10000 valid SNPs.\n",
      "The GRM computation is completed.\n",
      "Saving GRM...\n",
      "GRM has been saved in the file [data/grm/gcta.grm.bin]\n",
      "Number of SNPs in each pair of individuals has been saved in the file [data/grm/gcta.grm.N.bin]\n",
      "\n",
      "Analysis finished at 16:46:12 PST on Wed Nov 13 2024\n",
      "Overall computational time: 2.91 sec.\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "gcta64 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-grm \\\n",
    "    --out data/grm/gcta\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "07c663a0-b92a-448e-81b1-59541877e21c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\n",
      "* GCTA (Genome-wide Complex Trait Analysis)\n",
      "* Version v1.94.1 Mac\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang jian.yang@westlake.edu.cn\n",
      "* MIT License\n",
      "*******************************************************************\n",
      "Analysis started at 16:46:12 PST on Wed Nov 13 2024.\n",
      "Hostname: awlpdcg540mz\n",
      "\n",
      "Options: \n",
      " \n",
      "--grm data/grm/gcta \n",
      "--make-bK-sparse 0.05 \n",
      "--out data/grm/gcta-sparse \n",
      "\n",
      "Pruning the GRM to a sparse matrix with a cutoff of 0.050000...\n",
      "Total number of parts to be processed: 2\n",
      "    Saving 10000 individual IDs\n",
      "    Processing part 1\n",
      "Saving the sparse GRM (10018 pairs) to [data/grm/gcta-sparse.grm.sp]\n",
      "finished generating a sparse GRM Success:\n",
      "\n",
      "Analysis finished at 16:46:12 PST on Wed Nov 13 2024\n",
      "Overall computational time: 0.18 sec.\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "gcta64 \\\n",
    "    --grm data/grm/gcta \\\n",
    "    --make-bK-sparse 0.05 \\\n",
    "    --out data/grm/gcta-sparse\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "299d5c4a-3750-4ad1-8037-c144a15eb4ae",
   "metadata": {},
   "source": [
    "# GWAS\n",
    "\n",
    "## OLS univariate\n",
    "\n",
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0ec7a9c5-3db0-4433-9a44-3d00a3331f04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a6 M1 (18 Aug 2024)                 www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/ols_univariate/features/result.log.\n",
      "Options in effect:\n",
      "  --bfile data/simulated-phenotypes/Genotypes\n",
      "  --glm allow-no-covars\n",
      "  --out data/gwas/ols_univariate/features/result\n",
      "  --pheno data/simulated-phenotypes/Ysim_plink.txt\n",
      "\n",
      "Start time: Wed Nov 13 16:46:12 2024\n",
      "65536 MiB RAM detected; reserving 32768 MiB for main workspace.\n",
      "Using up to 16 threads (change this with --threads).\n",
      "10000 samples (0 females, 0 males, 10000 ambiguous; 10000 founders) loaded from\n",
      "data/simulated-phenotypes/Genotypes.fam.\n",
      "10000 variants loaded from data/simulated-phenotypes/Genotypes.bim.\n",
      "10 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-10: done.\n",
      "Results written to data/gwas/ols_univariate/features/result.<phenotype\n",
      "name>.glm.linear .\n",
      "End time: Wed Nov 13 16:46:12 2024\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "plink2 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --pheno data/simulated-phenotypes/Ysim_plink.txt \\\n",
    "    --glm allow-no-covars \\\n",
    "    --out data/gwas/ols_univariate/features/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e750c15f-7e05-4e49-9a9f-12cfe028e9d0",
   "metadata": {},
   "source": [
    "### Projections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c8df9887-8aac-4a95-b532-f4608e324966",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a6 M1 (18 Aug 2024)                 www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/ols_univariate/projections/result.log.\n",
      "Options in effect:\n",
      "  --bfile data/simulated-phenotypes/Genotypes\n",
      "  --glm allow-no-covars\n",
      "  --out data/gwas/ols_univariate/projections/result\n",
      "  --pheno data/projected-phenotypes/phenotypes.tsv\n",
      "\n",
      "Start time: Wed Nov 13 16:46:12 2024\n",
      "65536 MiB RAM detected; reserving 32768 MiB for main workspace.\n",
      "Using up to 16 threads (change this with --threads).\n",
      "10000 samples (0 females, 0 males, 10000 ambiguous; 10000 founders) loaded from\n",
      "data/simulated-phenotypes/Genotypes.fam.\n",
      "10000 variants loaded from data/simulated-phenotypes/Genotypes.bim.\n",
      "10 quantitative phenotypes loaded.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-10: done.\n",
      "Results written to data/gwas/ols_univariate/projections/result.<phenotype\n",
      "name>.glm.linear .\n",
      "End time: Wed Nov 13 16:46:12 2024\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "plink2 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --pheno data/projected-phenotypes/phenotypes.tsv \\\n",
    "    --glm allow-no-covars \\\n",
    "    --out data/gwas/ols_univariate/projections/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43d1d91f-432a-473b-af84-fa53cc880d7d",
   "metadata": {},
   "source": [
    "## OLS multivariate\n",
    "\n",
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ccf0963b-d482-4943-965b-48dd8aa367db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a6 M1 (18 Aug 2024)                 www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/ols_multivariate/features/result.log.\n",
      "Options in effect:\n",
      "  --bfile data/simulated-phenotypes/Genotypes\n",
      "  --covar data/simulated-phenotypes/Covs_plink.txt\n",
      "  --glm hide-covar\n",
      "  --out data/gwas/ols_multivariate/features/result\n",
      "  --pheno data/simulated-phenotypes/Ysim_plink.txt\n",
      "\n",
      "Start time: Wed Nov 13 16:46:12 2024\n",
      "65536 MiB RAM detected; reserving 32768 MiB for main workspace.\n",
      "Using up to 16 threads (change this with --threads).\n",
      "10000 samples (0 females, 0 males, 10000 ambiguous; 10000 founders) loaded from\n",
      "data/simulated-phenotypes/Genotypes.fam.\n",
      "10000 variants loaded from data/simulated-phenotypes/Genotypes.bim.\n",
      "10 quantitative phenotypes loaded.\n",
      "5 covariates loaded from data/simulated-phenotypes/Covs_plink.txt.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-10: done.\n",
      "Results written to data/gwas/ols_multivariate/features/result.<phenotype\n",
      "name>.glm.linear .\n",
      "End time: Wed Nov 13 16:46:13 2024\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "plink2 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --pheno data/simulated-phenotypes/Ysim_plink.txt \\\n",
    "    --covar data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --glm hide-covar \\\n",
    "    --out data/gwas/ols_multivariate/features/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d62d1a77-fa93-4e37-8155-60b962574254",
   "metadata": {},
   "source": [
    "### Projections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d6b079dc-ea8d-4ea5-b308-1ff2a579bf48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a6 M1 (18 Aug 2024)                 www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/gwas/ols_multivariate/projections/result.log.\n",
      "Options in effect:\n",
      "  --bfile data/simulated-phenotypes/Genotypes\n",
      "  --covar data/simulated-phenotypes/Covs_plink.txt\n",
      "  --glm hide-covar\n",
      "  --out data/gwas/ols_multivariate/projections/result\n",
      "  --pheno data/projected-phenotypes/phenotypes.tsv\n",
      "\n",
      "Start time: Wed Nov 13 16:46:13 2024\n",
      "65536 MiB RAM detected; reserving 32768 MiB for main workspace.\n",
      "Using up to 16 threads (change this with --threads).\n",
      "10000 samples (0 females, 0 males, 10000 ambiguous; 10000 founders) loaded from\n",
      "data/simulated-phenotypes/Genotypes.fam.\n",
      "10000 variants loaded from data/simulated-phenotypes/Genotypes.bim.\n",
      "10 quantitative phenotypes loaded.\n",
      "5 covariates loaded from data/simulated-phenotypes/Covs_plink.txt.\n",
      "Calculating allele frequencies... done.\n",
      "--glm linear regression on quantitative phenotypes #1-10: done.\n",
      "Results written to data/gwas/ols_multivariate/projections/result.<phenotype\n",
      "name>.glm.linear .\n",
      "End time: Wed Nov 13 16:46:13 2024\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "plink2 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --pheno data/projected-phenotypes/phenotypes.tsv \\\n",
    "    --covar data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --glm hide-covar \\\n",
    "    --out data/gwas/ols_multivariate/projections/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61197a40-0460-4034-970e-538586b7595a",
   "metadata": {},
   "source": [
    "## Regenie\n",
    "\n",
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c0920238-716c-4fba-91bc-f2ba4a65cba7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time: Thu Nov 14 00:46:15 2024\n",
      "\n",
      "              |===========================|\n",
      "              |      REGENIE v4.0.gz      |\n",
      "              |===========================|\n",
      "\n",
      "Copyright (c) 2020-2024 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "Compiled with Boost Iostream library.\n",
      "Compiled with HTSlib.\n",
      "Using Intel MKL with Eigen.\n",
      "\n",
      "Log of output saved in file : data/gwas/regenie/features/result.log\n",
      "\n",
      "Options in effect:\n",
      "  --step 1 \\\n",
      "  --bed data/simulated-phenotypes/Genotypes \\\n",
      "  --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
      "  --phenoFile data/simulated-phenotypes/Ysim_plink_names.txt \\\n",
      "  --bsize 100 \\\n",
      "  --qt \\\n",
      "  --out data/gwas/regenie/features/result\n",
      "\n",
      "Fitting null model\n",
      " * bim              : [data/simulated-phenotypes/Genotypes.bim] n_snps = 10000\n",
      " * fam              : [data/simulated-phenotypes/Genotypes.fam] n_samples = 10000\n",
      " * bed              : [data/simulated-phenotypes/Genotypes.bed]\n",
      " * phenotypes       : [data/simulated-phenotypes/Ysim_plink_names.txt] n_pheno = 10\n",
      "   -keeping and mean-imputing missing observations (done for each trait)\n",
      "   -number of phenotyped individuals  = 10000\n",
      " * covariates       : [data/simulated-phenotypes/Covs_plink.txt] n_cov = 5\n",
      "   -number of individuals with covariate data = 10000\n",
      " * number of individuals used in analysis = 10000\n",
      " * number of observations for each trait:\n",
      "   - 'feature_01': 10000 observations\n",
      "   - 'feature_02': 10000 observations\n",
      "   - 'feature_03': 10000 observations\n",
      "   - 'feature_04': 10000 observations\n",
      "   - 'feature_05': 10000 observations\n",
      "   - 'feature_06': 10000 observations\n",
      "   - 'feature_07': 10000 observations\n",
      "   - 'feature_08': 10000 observations\n",
      "   - 'feature_09': 10000 observations\n",
      "   - 'feature_10': 10000 observations\n",
      "   -residualizing and scaling phenotypes...done (1ms) \n",
      " * # threads        : [15]\n",
      " * block size       : [100]\n",
      " * # blocks         : [100] for 10000 variants\n",
      " * # CV folds       : [5]\n",
      " * ridge data_l0    : [ 5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * ridge data_l1    : [ 5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * approximate memory usage : 416MB\n",
      " * setting memory...done\n",
      "\n",
      "Chromosome 1\n",
      " block [1] : 100 snps  (17ms) \n",
      "   -residualizing and scaling genotypes...done (4ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (22ms) \n",
      " block [2] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (3ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [3] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [4] : 100 snps  (23ms) \n",
      "   -residualizing and scaling genotypes...done (3ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [5] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [6] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [7] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [8] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [9] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [10] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [11] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [12] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [13] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [14] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [15] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [16] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [17] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [18] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [19] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [20] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [21] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [22] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [23] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [24] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [25] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [26] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [27] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [28] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [29] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [30] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [31] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [32] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [33] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [34] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [35] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [36] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [37] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [38] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [39] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [40] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [41] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [42] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [43] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [44] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [45] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [46] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [47] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [48] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (14ms) \n",
      " block [49] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [50] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [51] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [52] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (14ms) \n",
      " block [53] : 100 snps  (5ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [54] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [55] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [56] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [57] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [58] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [59] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [60] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [61] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [62] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [63] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [64] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [65] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (32ms) \n",
      " block [66] : 100 snps  (20ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [67] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [68] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [69] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [70] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [71] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [72] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [73] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [74] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [75] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [76] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [77] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [78] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [79] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [80] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [81] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [82] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [83] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [84] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [85] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [86] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [87] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [88] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [89] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [90] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [91] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [92] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [93] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [94] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [95] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [96] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [97] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [98] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [99] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [100] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      "\n",
      " Level 1 ridge...\n",
      "   -on phenotype 1 (feature_01)...done (403ms) \n",
      "   -on phenotype 2 (feature_02)...done (384ms) \n",
      "   -on phenotype 3 (feature_03)...done (375ms) \n",
      "   -on phenotype 4 (feature_04)...done (380ms) \n",
      "   -on phenotype 5 (feature_05)...done (375ms) \n",
      "   -on phenotype 6 (feature_06)...done (377ms) \n",
      "   -on phenotype 7 (feature_07)...done (374ms) \n",
      "   -on phenotype 8 (feature_08)...done (374ms) \n",
      "   -on phenotype 9 (feature_09)...done (375ms) \n",
      "   -on phenotype 10 (feature_10)...done (371ms) \n",
      "\n",
      "Output\n",
      "------\n",
      "phenotype 1 (feature_01) : \n",
      "  0.01  : Rsq = 0.0189239, MSE = 0.98332\n",
      "  0.25  : Rsq = 0.018925, MSE = 0.982424\n",
      "  0.5   : Rsq = 0.019874, MSE = 0.981636\n",
      "  0.75  : Rsq = 0.0217418, MSE = 0.979644<- min value\n",
      "  0.99  : Rsq = 0.0237075, MSE = 0.982091\n",
      "  * making predictions...writing LOCO predictions...done (66ms) \n",
      "\n",
      "phenotype 2 (feature_02) : \n",
      "  0.01  : Rsq = 0.236413, MSE = 0.83006\n",
      "  0.25  : Rsq = 0.236105, MSE = 0.763451\n",
      "  0.5   : Rsq = 0.236615, MSE = 0.763177\n",
      "  0.75  : Rsq = 0.237428, MSE = 0.762528<- min value\n",
      "  0.99  : Rsq = 0.233683, MSE = 0.767384\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 3 (feature_03) : \n",
      "  0.01  : Rsq = 0.249361, MSE = 0.818247\n",
      "  0.25  : Rsq = 0.247927, MSE = 0.751636<- min value\n",
      "  0.5   : Rsq = 0.248057, MSE = 0.751745\n",
      "  0.75  : Rsq = 0.248209, MSE = 0.751775\n",
      "  0.99  : Rsq = 0.243674, MSE = 0.757272\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 4 (feature_04) : \n",
      "  0.01  : Rsq = 0.350888, MSE = 0.752361\n",
      "  0.25  : Rsq = 0.351077, MSE = 0.648573<- min value\n",
      "  0.5   : Rsq = 0.351014, MSE = 0.648674\n",
      "  0.75  : Rsq = 0.35067, MSE = 0.649166\n",
      "  0.99  : Rsq = 0.34427, MSE = 0.656292\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 5 (feature_05) : \n",
      "  0.01  : Rsq = 0.329195, MSE = 0.763748\n",
      "  0.25  : Rsq = 0.328711, MSE = 0.670909<- min value\n",
      "  0.5   : Rsq = 0.328744, MSE = 0.670938\n",
      "  0.75  : Rsq = 0.328612, MSE = 0.671209\n",
      "  0.99  : Rsq = 0.322607, MSE = 0.677984\n",
      "  * making predictions...writing LOCO predictions...done (64ms) \n",
      "\n",
      "phenotype 6 (feature_06) : \n",
      "  0.01  : Rsq = 0.0706465, MSE = 0.945975\n",
      "  0.25  : Rsq = 0.0705834, MSE = 0.929446\n",
      "  0.5   : Rsq = 0.0713099, MSE = 0.929058\n",
      "  0.75  : Rsq = 0.0726141, MSE = 0.927867<- min value\n",
      "  0.99  : Rsq = 0.071595, MSE = 0.931089\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 7 (feature_07) : \n",
      "  0.01  : Rsq = 0.137263, MSE = 0.898991\n",
      "  0.25  : Rsq = 0.136705, MSE = 0.86301\n",
      "  0.5   : Rsq = 0.137442, MSE = 0.862632\n",
      "  0.75  : Rsq = 0.138717, MSE = 0.86154<- min value\n",
      "  0.99  : Rsq = 0.135935, MSE = 0.86643\n",
      "  * making predictions...writing LOCO predictions...done (64ms) \n",
      "\n",
      "phenotype 8 (feature_08) : \n",
      "  0.01  : Rsq = 0.285399, MSE = 0.798451\n",
      "  0.25  : Rsq = 0.286135, MSE = 0.713453\n",
      "  0.5   : Rsq = 0.286542, MSE = 0.713117\n",
      "  0.75  : Rsq = 0.287162, MSE = 0.712617<- min value\n",
      "  0.99  : Rsq = 0.283724, MSE = 0.716718\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 9 (feature_09) : \n",
      "  0.01  : Rsq = 0.119667, MSE = 0.911429\n",
      "  0.25  : Rsq = 0.118679, MSE = 0.881071<- min value\n",
      "  0.5   : Rsq = 0.118613, MSE = 0.881538\n",
      "  0.75  : Rsq = 0.118262, MSE = 0.882173\n",
      "  0.99  : Rsq = 0.111919, MSE = 0.890744\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 10 (feature_10) : \n",
      "  0.01  : Rsq = 0.202909, MSE = 0.853554\n",
      "  0.25  : Rsq = 0.202312, MSE = 0.797249\n",
      "  0.5   : Rsq = 0.202552, MSE = 0.797286\n",
      "  0.75  : Rsq = 0.202845, MSE = 0.797188<- min value\n",
      "  0.99  : Rsq = 0.198132, MSE = 0.803202\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "List of blup files written to: [data/gwas/regenie/features/result_pred.list]\n",
      "\n",
      "Elapsed time : 7.96641s\n",
      "End time: Thu Nov 14 00:46:23 2024\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    ghcr.io/rgcgithub/regenie/regenie:v4.0.gz regenie \\\n",
    "    --step 1 \\\n",
    "    --bed data/simulated-phenotypes/Genotypes \\\n",
    "    --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --phenoFile data/simulated-phenotypes/Ysim_plink_names.txt \\\n",
    "    --bsize 100 \\\n",
    "    --qt \\\n",
    "    --out data/gwas/regenie/features/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ea785d2a-a13a-4b97-a73a-db1f4e864058",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time: Thu Nov 14 00:46:23 2024\n",
      "\n",
      "              |===========================|\n",
      "              |      REGENIE v4.0.gz      |\n",
      "              |===========================|\n",
      "\n",
      "Copyright (c) 2020-2024 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "Compiled with Boost Iostream library.\n",
      "Compiled with HTSlib.\n",
      "Using Intel MKL with Eigen.\n",
      "\n",
      "Log of output saved in file : data/gwas/regenie/features/result.log\n",
      "\n",
      "Options in effect:\n",
      "  --step 2 \\\n",
      "  --bed data/simulated-phenotypes/Genotypes \\\n",
      "  --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
      "  --phenoFile data/simulated-phenotypes/Ysim_plink_names.txt \\\n",
      "  --bsize 100 \\\n",
      "  --qt \\\n",
      "  --pred data/gwas/regenie/features/result_pred.list \\\n",
      "  --out data/gwas/regenie/features/result\n",
      "\n",
      "Association testing mode with fast multithreading using OpenMP\n",
      " * bim              : [data/simulated-phenotypes/Genotypes.bim] n_snps = 10000\n",
      " * fam              : [data/simulated-phenotypes/Genotypes.fam] n_samples = 10000\n",
      " * bed              : [data/simulated-phenotypes/Genotypes.bed]\n",
      " * phenotypes       : [data/simulated-phenotypes/Ysim_plink_names.txt] n_pheno = 10\n",
      "   -number of phenotyped individuals  = 10000\n",
      " * covariates       : [data/simulated-phenotypes/Covs_plink.txt] n_cov = 5\n",
      "   -number of individuals with covariate data = 10000\n",
      " * number of individuals used in analysis = 10000\n",
      " * number of observations for each trait:\n",
      "   - 'feature_01': 10000 observations\n",
      "   - 'feature_02': 10000 observations\n",
      "   - 'feature_03': 10000 observations\n",
      "   - 'feature_04': 10000 observations\n",
      "   - 'feature_05': 10000 observations\n",
      "   - 'feature_06': 10000 observations\n",
      "   - 'feature_07': 10000 observations\n",
      "   - 'feature_08': 10000 observations\n",
      "   - 'feature_09': 10000 observations\n",
      "   - 'feature_10': 10000 observations\n",
      " * LOCO predictions : [data/gwas/regenie/features/result_pred.list]\n",
      "   -file [/data/gwas/regenie/features/result_1.loco] for phenotype 'feature_01'\n",
      "   -file [/data/gwas/regenie/features/result_2.loco] for phenotype 'feature_02'\n",
      "   -file [/data/gwas/regenie/features/result_3.loco] for phenotype 'feature_03'\n",
      "   -file [/data/gwas/regenie/features/result_4.loco] for phenotype 'feature_04'\n",
      "   -file [/data/gwas/regenie/features/result_5.loco] for phenotype 'feature_05'\n",
      "   -file [/data/gwas/regenie/features/result_6.loco] for phenotype 'feature_06'\n",
      "   -file [/data/gwas/regenie/features/result_7.loco] for phenotype 'feature_07'\n",
      "   -file [/data/gwas/regenie/features/result_8.loco] for phenotype 'feature_08'\n",
      "   -file [/data/gwas/regenie/features/result_9.loco] for phenotype 'feature_09'\n",
      "   -file [/data/gwas/regenie/features/result_10.loco] for phenotype 'feature_10'\n",
      "   -residualizing and scaling phenotypes...done (1ms) \n",
      " * # threads        : [15]\n",
      " * block size       : [100]\n",
      " * # blocks         : [100]\n",
      " * approximate memory usage : 22MB\n",
      " * using minimum MAC of 5 (variants with lower MAC are ignored)\n",
      "\n",
      "Chromosome 1 [100 blocks in total]\n",
      "   -reading loco predictions for the chromosome...done (201ms) \n",
      " block [1/100] : done (75ms) \n",
      " block [2/100] : done (7ms) \n",
      " block [3/100] : done (5ms) \n",
      " block [4/100] : done (5ms) \n",
      " block [5/100] : done (6ms) \n",
      " block [6/100] : done (5ms) \n",
      " block [7/100] : done (5ms) \n",
      " block [8/100] : done (5ms) \n",
      " block [9/100] : done (5ms) \n",
      " block [10/100] : done (5ms) \n",
      " block [11/100] : done (5ms) \n",
      " block [12/100] : done (5ms) \n",
      " block [13/100] : done (5ms) \n",
      " block [14/100] : done (5ms) \n",
      " block [15/100] : done (6ms) \n",
      " block [16/100] : done (5ms) \n",
      " block [17/100] : done (5ms) \n",
      " block [18/100] : done (5ms) \n",
      " block [19/100] : done (5ms) \n",
      " block [20/100] : done (5ms) \n",
      " block [21/100] : done (5ms) \n",
      " block [22/100] : done (5ms) \n",
      " block [23/100] : done (5ms) \n",
      " block [24/100] : done (5ms) \n",
      " block [25/100] : done (5ms) \n",
      " block [26/100] : done (5ms) \n",
      " block [27/100] : done (5ms) \n",
      " block [28/100] : done (5ms) \n",
      " block [29/100] : done (5ms) \n",
      " block [30/100] : done (5ms) \n",
      " block [31/100] : done (5ms) \n",
      " block [32/100] : done (5ms) \n",
      " block [33/100] : done (5ms) \n",
      " block [34/100] : done (5ms) \n",
      " block [35/100] : done (5ms) \n",
      " block [36/100] : done (5ms) \n",
      " block [37/100] : done (5ms) \n",
      " block [38/100] : done (5ms) \n",
      " block [39/100] : done (5ms) \n",
      " block [40/100] : done (5ms) \n",
      " block [41/100] : done (5ms) \n",
      " block [42/100] : done (5ms) \n",
      " block [43/100] : done (5ms) \n",
      " block [44/100] : done (5ms) \n",
      " block [45/100] : done (5ms) \n",
      " block [46/100] : done (5ms) \n",
      " block [47/100] : done (5ms) \n",
      " block [48/100] : done (5ms) \n",
      " block [49/100] : done (5ms) \n",
      " block [50/100] : done (5ms) \n",
      " block [51/100] : done (5ms) \n",
      " block [52/100] : done (5ms) \n",
      " block [53/100] : done (5ms) \n",
      " block [54/100] : done (5ms) \n",
      " block [55/100] : done (5ms) \n",
      " block [56/100] : done (5ms) \n",
      " block [57/100] : done (5ms) \n",
      " block [58/100] : done (5ms) \n",
      " block [59/100] : done (5ms) \n",
      " block [60/100] : done (5ms) \n",
      " block [61/100] : done (5ms) \n",
      " block [62/100] : done (5ms) \n",
      " block [63/100] : done (5ms) \n",
      " block [64/100] : done (5ms) \n",
      " block [65/100] : done (5ms) \n",
      " block [66/100] : done (5ms) \n",
      " block [67/100] : done (5ms) \n",
      " block [68/100] : done (5ms) \n",
      " block [69/100] : done (5ms) \n",
      " block [70/100] : done (5ms) \n",
      " block [71/100] : done (5ms) \n",
      " block [72/100] : done (5ms) \n",
      " block [73/100] : done (5ms) \n",
      " block [74/100] : done (5ms) \n",
      " block [75/100] : done (5ms) \n",
      " block [76/100] : done (5ms) \n",
      " block [77/100] : done (5ms) \n",
      " block [78/100] : done (5ms) \n",
      " block [79/100] : done (6ms) \n",
      " block [80/100] : done (5ms) \n",
      " block [81/100] : done (5ms) \n",
      " block [82/100] : done (5ms) \n",
      " block [83/100] : done (5ms) \n",
      " block [84/100] : done (5ms) \n",
      " block [85/100] : done (5ms) \n",
      " block [86/100] : done (5ms) \n",
      " block [87/100] : done (5ms) \n",
      " block [88/100] : done (5ms) \n",
      " block [89/100] : done (5ms) \n",
      " block [90/100] : done (5ms) \n",
      " block [91/100] : done (5ms) \n",
      " block [92/100] : done (5ms) \n",
      " block [93/100] : done (5ms) \n",
      " block [94/100] : done (5ms) \n",
      " block [95/100] : done (5ms) \n",
      " block [96/100] : done (6ms) \n",
      " block [97/100] : done (5ms) \n",
      " block [98/100] : done (5ms) \n",
      " block [99/100] : done (5ms) \n",
      " block [100/100] : done (5ms) \n",
      "\n",
      "Association results stored separately for each trait in files : \n",
      "* [data/gwas/regenie/features/result_feature_01.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_02.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_03.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_04.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_05.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_06.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_07.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_08.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_09.regenie]\n",
      "* [data/gwas/regenie/features/result_feature_10.regenie]\n",
      "\n",
      "Number of ignored tests due to low MAC : 0\n",
      "\n",
      "Elapsed time : 1.56612s\n",
      "End time: Thu Nov 14 00:46:25 2024\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    ghcr.io/rgcgithub/regenie/regenie:v4.0.gz regenie \\\n",
    "    --step 2 \\\n",
    "    --bed data/simulated-phenotypes/Genotypes \\\n",
    "    --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --phenoFile data/simulated-phenotypes/Ysim_plink_names.txt \\\n",
    "    --bsize 100 \\\n",
    "    --qt \\\n",
    "    --pred data/gwas/regenie/features/result_pred.list \\\n",
    "    --out data/gwas/regenie/features/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3826ed25-0730-4458-a778-653f92461657",
   "metadata": {},
   "source": [
    "### Projections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "69fe0760-94e7-4aa5-b926-d54fcaaef270",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time: Thu Nov 14 00:46:25 2024\n",
      "\n",
      "              |===========================|\n",
      "              |      REGENIE v4.0.gz      |\n",
      "              |===========================|\n",
      "\n",
      "Copyright (c) 2020-2024 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "Compiled with Boost Iostream library.\n",
      "Compiled with HTSlib.\n",
      "Using Intel MKL with Eigen.\n",
      "\n",
      "Log of output saved in file : data/gwas/regenie/projections/result.log\n",
      "\n",
      "Options in effect:\n",
      "  --step 1 \\\n",
      "  --bed data/simulated-phenotypes/Genotypes \\\n",
      "  --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
      "  --phenoFile data/projected-phenotypes/phenotypes.tsv \\\n",
      "  --bsize 100 \\\n",
      "  --qt \\\n",
      "  --out data/gwas/regenie/projections/result\n",
      "\n",
      "Fitting null model\n",
      " * bim              : [data/simulated-phenotypes/Genotypes.bim] n_snps = 10000\n",
      " * fam              : [data/simulated-phenotypes/Genotypes.fam] n_samples = 10000\n",
      " * bed              : [data/simulated-phenotypes/Genotypes.bed]\n",
      " * phenotypes       : [data/projected-phenotypes/phenotypes.tsv] n_pheno = 10\n",
      "   -keeping and mean-imputing missing observations (done for each trait)\n",
      "   -number of phenotyped individuals  = 10000\n",
      " * covariates       : [data/simulated-phenotypes/Covs_plink.txt] n_cov = 5\n",
      "   -number of individuals with covariate data = 10000\n",
      " * number of individuals used in analysis = 10000\n",
      " * number of observations for each trait:\n",
      "   - 'projection_001': 10000 observations\n",
      "   - 'projection_002': 10000 observations\n",
      "   - 'projection_003': 10000 observations\n",
      "   - 'projection_004': 10000 observations\n",
      "   - 'projection_005': 10000 observations\n",
      "   - 'projection_006': 10000 observations\n",
      "   - 'projection_007': 10000 observations\n",
      "   - 'projection_008': 10000 observations\n",
      "   - 'projection_009': 10000 observations\n",
      "   - 'projection_010': 10000 observations\n",
      "   -residualizing and scaling phenotypes...done (1ms) \n",
      " * # threads        : [15]\n",
      " * block size       : [100]\n",
      " * # blocks         : [100] for 10000 variants\n",
      " * # CV folds       : [5]\n",
      " * ridge data_l0    : [ 5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * ridge data_l1    : [ 5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * approximate memory usage : 416MB\n",
      " * setting memory...done\n",
      "\n",
      "Chromosome 1\n",
      " block [1] : 100 snps  (18ms) \n",
      "   -residualizing and scaling genotypes...done (3ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (22ms) \n",
      " block [2] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [3] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [4] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [5] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [6] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [7] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [8] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [9] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [10] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [11] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [12] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [13] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [14] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [15] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [16] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [17] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [18] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [19] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [20] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [21] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [22] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [23] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [24] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [25] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [26] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [27] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [28] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [29] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [30] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [31] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [32] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [33] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [34] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [35] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [36] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [37] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [38] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [39] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [40] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [41] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [42] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [43] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [44] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [45] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [46] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [47] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [48] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [49] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [50] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [51] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [52] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [53] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (3ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [54] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [55] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [56] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [57] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [58] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [59] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [60] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [61] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [62] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [63] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [64] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (17ms) \n",
      " block [65] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [66] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [67] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [68] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [69] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (15ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [70] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [71] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [72] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [73] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [74] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [75] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [76] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [77] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [78] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [79] : 100 snps  (4ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [80] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [81] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [82] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [83] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [84] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [85] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [86] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [87] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (11ms) \n",
      " block [88] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [89] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [90] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [91] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [92] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [93] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [94] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [95] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [96] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [97] : 100 snps  (3ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (11ms) \n",
      "   -calc level 0 ridge...done (13ms) \n",
      " block [98] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (12ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [99] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      " block [100] : 100 snps  (2ms) \n",
      "   -residualizing and scaling genotypes...done (2ms) \n",
      "   -calc working matrices...done (10ms) \n",
      "   -calc level 0 ridge...done (12ms) \n",
      "\n",
      " Level 1 ridge...\n",
      "   -on phenotype 1 (projection_001)...done (399ms) \n",
      "   -on phenotype 2 (projection_002)...done (388ms) \n",
      "   -on phenotype 3 (projection_003)...done (393ms) \n",
      "   -on phenotype 4 (projection_004)...done (396ms) \n",
      "   -on phenotype 5 (projection_005)...done (395ms) \n",
      "   -on phenotype 6 (projection_006)...done (386ms) \n",
      "   -on phenotype 7 (projection_007)...done (391ms) \n",
      "   -on phenotype 8 (projection_008)...done (389ms) \n",
      "   -on phenotype 9 (projection_009)...done (372ms) \n",
      "   -on phenotype 10 (projection_010)...done (375ms) \n",
      "\n",
      "Output\n",
      "------\n",
      "phenotype 1 (projection_001) : \n",
      "  0.01  : Rsq = 0.232675, MSE = 0.830827\n",
      "  0.25  : Rsq = 0.23125, MSE = 0.768306<- min value\n",
      "  0.5   : Rsq = 0.231256, MSE = 0.768541\n",
      "  0.75  : Rsq = 0.231086, MSE = 0.768912\n",
      "  0.99  : Rsq = 0.224679, MSE = 0.776638\n",
      "  * making predictions...writing LOCO predictions...done (64ms) \n",
      "\n",
      "phenotype 2 (projection_002) : \n",
      "  0.01  : Rsq = 0.273858, MSE = 0.805263\n",
      "  0.25  : Rsq = 0.272757, MSE = 0.726813<- min value\n",
      "  0.5   : Rsq = 0.272647, MSE = 0.727185\n",
      "  0.75  : Rsq = 0.272254, MSE = 0.727807\n",
      "  0.99  : Rsq = 0.265577, MSE = 0.73563\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 3 (projection_003) : \n",
      "  0.01  : Rsq = 0.254056, MSE = 0.818159\n",
      "  0.25  : Rsq = 0.25341, MSE = 0.746145\n",
      "  0.5   : Rsq = 0.253723, MSE = 0.746028\n",
      "  0.75  : Rsq = 0.254205, MSE = 0.74571<- min value\n",
      "  0.99  : Rsq = 0.249909, MSE = 0.751004\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 4 (projection_004) : \n",
      "  0.01  : Rsq = 0.328904, MSE = 0.767533\n",
      "  0.25  : Rsq = 0.329779, MSE = 0.66986\n",
      "  0.5   : Rsq = 0.329967, MSE = 0.669696\n",
      "  0.75  : Rsq = 0.33014, MSE = 0.669644<- min value\n",
      "  0.99  : Rsq = 0.325516, MSE = 0.67488\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 5 (projection_005) : \n",
      "  0.01  : Rsq = 0.191626, MSE = 0.859852\n",
      "  0.25  : Rsq = 0.190558, MSE = 0.809046\n",
      "  0.5   : Rsq = 0.190825, MSE = 0.8091\n",
      "  0.75  : Rsq = 0.191215, MSE = 0.808912<- min value\n",
      "  0.99  : Rsq = 0.186936, MSE = 0.814511\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 6 (projection_006) : \n",
      "  0.01  : Rsq = 0.165066, MSE = 0.880383\n",
      "  0.25  : Rsq = 0.164387, MSE = 0.835232\n",
      "  0.5   : Rsq = 0.16483, MSE = 0.835121\n",
      "  0.75  : Rsq = 0.165544, MSE = 0.8346<- min value\n",
      "  0.99  : Rsq = 0.16179, MSE = 0.839942\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 7 (projection_007) : \n",
      "  0.01  : Rsq = 0.18314, MSE = 0.866872\n",
      "  0.25  : Rsq = 0.181894, MSE = 0.817709<- min value\n",
      "  0.5   : Rsq = 0.181933, MSE = 0.818005\n",
      "  0.75  : Rsq = 0.181807, MSE = 0.818366\n",
      "  0.99  : Rsq = 0.175493, MSE = 0.826284\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 8 (projection_008) : \n",
      "  0.01  : Rsq = 0.288633, MSE = 0.794672\n",
      "  0.25  : Rsq = 0.288217, MSE = 0.711359<- min value\n",
      "  0.5   : Rsq = 0.288224, MSE = 0.711494\n",
      "  0.75  : Rsq = 0.288088, MSE = 0.711803\n",
      "  0.99  : Rsq = 0.282536, MSE = 0.718256\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "phenotype 9 (projection_009) : \n",
      "  0.01  : Rsq = 0.334404, MSE = 0.763088\n",
      "  0.25  : Rsq = 0.33416, MSE = 0.665452\n",
      "  0.5   : Rsq = 0.334427, MSE = 0.665296\n",
      "  0.75  : Rsq = 0.334825, MSE = 0.665052<- min value\n",
      "  0.99  : Rsq = 0.330771, MSE = 0.669837\n",
      "  * making predictions...writing LOCO predictions...done (63ms) \n",
      "\n",
      "phenotype 10 (projection_010) : \n",
      "  0.01  : Rsq = 0.143824, MSE = 0.893287\n",
      "  0.25  : Rsq = 0.142443, MSE = 0.857212\n",
      "  0.5   : Rsq = 0.142772, MSE = 0.85721\n",
      "  0.75  : Rsq = 0.143311, MSE = 0.856865<- min value\n",
      "  0.99  : Rsq = 0.139484, MSE = 0.862393\n",
      "  * making predictions...writing LOCO predictions...done (62ms) \n",
      "\n",
      "List of blup files written to: [data/gwas/regenie/projections/result_pred.list]\n",
      "\n",
      "Elapsed time : 7.89208s\n",
      "End time: Thu Nov 14 00:46:33 2024\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    ghcr.io/rgcgithub/regenie/regenie:v4.0.gz regenie \\\n",
    "    --step 1 \\\n",
    "    --bed data/simulated-phenotypes/Genotypes \\\n",
    "    --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --phenoFile data/projected-phenotypes/phenotypes.tsv \\\n",
    "    --bsize 100 \\\n",
    "    --qt \\\n",
    "    --out data/gwas/regenie/projections/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd7a812c-f795-43df-b178-f57992aba8e8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time: Thu Nov 14 00:46:33 2024\n",
      "\n",
      "              |===========================|\n",
      "              |      REGENIE v4.0.gz      |\n",
      "              |===========================|\n",
      "\n",
      "Copyright (c) 2020-2024 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "Compiled with Boost Iostream library.\n",
      "Compiled with HTSlib.\n",
      "Using Intel MKL with Eigen.\n",
      "\n",
      "Log of output saved in file : data/gwas/regenie/projections/result.log\n",
      "\n",
      "Options in effect:\n",
      "  --step 2 \\\n",
      "  --bed data/simulated-phenotypes/Genotypes \\\n",
      "  --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
      "  --phenoFile data/projected-phenotypes/phenotypes.tsv \\\n",
      "  --bsize 100 \\\n",
      "  --qt \\\n",
      "  --pred data/gwas/regenie/projections/result_pred.list \\\n",
      "  --out data/gwas/regenie/projections/result\n",
      "\n",
      "Association testing mode with fast multithreading using OpenMP\n",
      " * bim              : [data/simulated-phenotypes/Genotypes.bim] n_snps = 10000\n",
      " * fam              : [data/simulated-phenotypes/Genotypes.fam] n_samples = 10000\n",
      " * bed              : [data/simulated-phenotypes/Genotypes.bed]\n",
      " * phenotypes       : [data/projected-phenotypes/phenotypes.tsv] n_pheno = 10\n",
      "   -number of phenotyped individuals  = 10000\n",
      " * covariates       : [data/simulated-phenotypes/Covs_plink.txt] n_cov = 5\n",
      "   -number of individuals with covariate data = 10000\n",
      " * number of individuals used in analysis = 10000\n",
      " * number of observations for each trait:\n",
      "   - 'projection_001': 10000 observations\n",
      "   - 'projection_002': 10000 observations\n",
      "   - 'projection_003': 10000 observations\n",
      "   - 'projection_004': 10000 observations\n",
      "   - 'projection_005': 10000 observations\n",
      "   - 'projection_006': 10000 observations\n",
      "   - 'projection_007': 10000 observations\n",
      "   - 'projection_008': 10000 observations\n",
      "   - 'projection_009': 10000 observations\n",
      "   - 'projection_010': 10000 observations\n",
      " * LOCO predictions : [data/gwas/regenie/projections/result_pred.list]\n",
      "   -file [/data/gwas/regenie/projections/result_1.loco] for phenotype 'projection_001'\n",
      "   -file [/data/gwas/regenie/projections/result_2.loco] for phenotype 'projection_002'\n",
      "   -file [/data/gwas/regenie/projections/result_3.loco] for phenotype 'projection_003'\n",
      "   -file [/data/gwas/regenie/projections/result_4.loco] for phenotype 'projection_004'\n",
      "   -file [/data/gwas/regenie/projections/result_5.loco] for phenotype 'projection_005'\n",
      "   -file [/data/gwas/regenie/projections/result_6.loco] for phenotype 'projection_006'\n",
      "   -file [/data/gwas/regenie/projections/result_7.loco] for phenotype 'projection_007'\n",
      "   -file [/data/gwas/regenie/projections/result_8.loco] for phenotype 'projection_008'\n",
      "   -file [/data/gwas/regenie/projections/result_9.loco] for phenotype 'projection_009'\n",
      "   -file [/data/gwas/regenie/projections/result_10.loco] for phenotype 'projection_010'\n",
      "   -residualizing and scaling phenotypes...done (1ms) \n",
      " * # threads        : [15]\n",
      " * block size       : [100]\n",
      " * # blocks         : [100]\n",
      " * approximate memory usage : 22MB\n",
      " * using minimum MAC of 5 (variants with lower MAC are ignored)\n",
      "\n",
      "Chromosome 1 [100 blocks in total]\n",
      "   -reading loco predictions for the chromosome...done (199ms) \n",
      " block [1/100] : done (75ms) \n",
      " block [2/100] : done (6ms) \n",
      " block [3/100] : done (5ms) \n",
      " block [4/100] : done (5ms) \n",
      " block [5/100] : done (5ms) \n",
      " block [6/100] : done (5ms) \n",
      " block [7/100] : done (5ms) \n",
      " block [8/100] : done (5ms) \n",
      " block [9/100] : done (5ms) \n",
      " block [10/100] : done (5ms) \n",
      " block [11/100] : done (5ms) \n",
      " block [12/100] : done (5ms) \n",
      " block [13/100] : done (5ms) \n",
      " block [14/100] : done (5ms) \n",
      " block [15/100] : done (5ms) \n",
      " block [16/100] : done (5ms) \n",
      " block [17/100] : done (5ms) \n",
      " block [18/100] : done (5ms) \n",
      " block [19/100] : done (5ms) \n",
      " block [20/100] : done (5ms) \n",
      " block [21/100] : done (5ms) \n",
      " block [22/100] : done (5ms) \n",
      " block [23/100] : done (5ms) \n",
      " block [24/100] : done (5ms) \n",
      " block [25/100] : done (5ms) \n",
      " block [26/100] : done (5ms) \n",
      " block [27/100] : done (5ms) \n",
      " block [28/100] : done (5ms) \n",
      " block [29/100] : done (5ms) \n",
      " block [30/100] : done (5ms) \n",
      " block [31/100] : done (5ms) \n",
      " block [32/100] : done (5ms) \n",
      " block [33/100] : done (5ms) \n",
      " block [34/100] : done (5ms) \n",
      " block [35/100] : done (5ms) \n",
      " block [36/100] : done (5ms) \n",
      " block [37/100] : done (5ms) \n",
      " block [38/100] : done (5ms) \n",
      " block [39/100] : done (5ms) \n",
      " block [40/100] : done (5ms) \n",
      " block [41/100] : done (5ms) \n",
      " block [42/100] : done (5ms) \n",
      " block [43/100] : done (5ms) \n",
      " block [44/100] : done (5ms) \n",
      " block [45/100] : done (5ms) \n",
      " block [46/100] : done (5ms) \n",
      " block [47/100] : done (5ms) \n",
      " block [48/100] : done (5ms) \n",
      " block [49/100] : done (5ms) \n",
      " block [50/100] : done (5ms) \n",
      " block [51/100] : done (5ms) \n",
      " block [52/100] : done (5ms) \n",
      " block [53/100] : done (5ms) \n",
      " block [54/100] : done (5ms) \n",
      " block [55/100] : done (5ms) \n",
      " block [56/100] : done (5ms) \n",
      " block [57/100] : done (5ms) \n",
      " block [58/100] : done (5ms) \n",
      " block [59/100] : done (5ms) \n",
      " block [60/100] : done (5ms) \n",
      " block [61/100] : done (5ms) \n",
      " block [62/100] : done (5ms) \n",
      " block [63/100] : done (5ms) \n",
      " block [64/100] : done (5ms) \n",
      " block [65/100] : done (5ms) \n",
      " block [66/100] : done (5ms) \n",
      " block [67/100] : done (5ms) \n",
      " block [68/100] : done (5ms) \n",
      " block [69/100] : done (5ms) \n",
      " block [70/100] : done (5ms) \n",
      " block [71/100] : done (5ms) \n",
      " block [72/100] : done (5ms) \n",
      " block [73/100] : done (5ms) \n",
      " block [74/100] : done (5ms) \n",
      " block [75/100] : done (5ms) \n",
      " block [76/100] : done (5ms) \n",
      " block [77/100] : done (5ms) \n",
      " block [78/100] : done (5ms) \n",
      " block [79/100] : done (5ms) \n",
      " block [80/100] : done (5ms) \n",
      " block [81/100] : done (5ms) \n",
      " block [82/100] : done (5ms) \n",
      " block [83/100] : done (5ms) \n",
      " block [84/100] : done (5ms) \n",
      " block [85/100] : done (5ms) \n",
      " block [86/100] : done (5ms) \n",
      " block [87/100] : done (5ms) \n",
      " block [88/100] : done (5ms) \n",
      " block [89/100] : done (5ms) \n",
      " block [90/100] : done (5ms) \n",
      " block [91/100] : done (5ms) \n",
      " block [92/100] : done (5ms) \n",
      " block [93/100] : done (5ms) \n",
      " block [94/100] : done (5ms) \n",
      " block [95/100] : done (5ms) \n",
      " block [96/100] : done (5ms) \n",
      " block [97/100] : done (5ms) \n",
      " block [98/100] : done (5ms) \n",
      " block [99/100] : done (5ms) \n",
      " block [100/100] : done (5ms) \n",
      "\n",
      "Association results stored separately for each trait in files : \n",
      "* [data/gwas/regenie/projections/result_projection_001.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_002.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_003.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_004.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_005.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_006.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_007.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_008.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_009.regenie]\n",
      "* [data/gwas/regenie/projections/result_projection_010.regenie]\n",
      "\n",
      "Number of ignored tests due to low MAC : 0\n",
      "\n",
      "Elapsed time : 1.55593s\n",
      "End time: Thu Nov 14 00:46:35 2024\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command = \"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    ghcr.io/rgcgithub/regenie/regenie:v4.0.gz regenie \\\n",
    "    --step 2 \\\n",
    "    --bed data/simulated-phenotypes/Genotypes \\\n",
    "    --covarFile data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --phenoFile data/projected-phenotypes/phenotypes.tsv \\\n",
    "    --bsize 100 \\\n",
    "    --qt \\\n",
    "    --pred data/gwas/regenie/projections/result_pred.list \\\n",
    "    --out data/gwas/regenie/projections/result\n",
    "\"\"\"\n",
    "result = subprocess.run(shlex.split(command))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3396017-bbdc-4760-9d0e-bf7e2c00c9ae",
   "metadata": {},
   "source": [
    "## SAIGE\n",
    "\n",
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fcda23e4-9352-42b7-a3c0-ebf9c9d00eb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3ec0ab555de344b2a548d8681d41ae21",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    wzhou88/saige:1.3.0 step1_fitNULLGLMM.R \\\n",
    "    --plinkFile=data/simulated-phenotypes/Genotypes \\\n",
    "    --useSparseGRMtoFitNULL=FALSE \\\n",
    "    --covarColList=sharedConfounder_norm1,sharedConfounder_norm2,sharedConfounder_norm3,independentConfounder_norm1,independentConfounder_norm2 \\\n",
    "    --sampleIDColinphenoFile=IID \\\n",
    "    --invNormalize=TRUE \\\n",
    "    --traitType=quantitative \\\n",
    "    --nThreads=16 \\\n",
    "    --IsOverwriteVarianceRatioFile=TRUE \\\n",
    "    --phenoCol=$PHENOTYPE \\\n",
    "    --phenoFile=data/simulated-phenotypes/Ysim_saige.txt \\\n",
    "    --outputPrefix=data/gwas/saige/features/$PHENOTYPE\n",
    "\"\"\")\n",
    "\n",
    "for feature_idx in tqdm.tnrange(1, n_features + 1):\n",
    "    command = command_template.substitute(PHENOTYPE=f\"feature_{feature_idx:02}\")\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d929403c-9b26-4760-9cf8-e1c073faa70d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fc1acc9db8ca41e3af7673d9bf4c423f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    wzhou88/saige:1.3.0 step2_SPAtests.R \\\n",
    "    --bedFile=data/simulated-phenotypes/Genotypes.bed \\\n",
    "    --bimFile=data/simulated-phenotypes/Genotypes.bim \\\n",
    "    --famFile=data/simulated-phenotypes/Genotypes.fam \\\n",
    "    --LOCO=FALSE \\\n",
    "    --GMMATmodelFile=data/gwas/saige/features/$PHENOTYPE.rda \\\n",
    "    --varianceRatioFile=data/gwas/saige/features/$PHENOTYPE.varianceRatio.txt \\\n",
    "    --SAIGEOutputFile=data/gwas/saige/features/$PHENOTYPE.txt\n",
    "\"\"\")\n",
    "\n",
    "for feature_idx in tqdm.tnrange(1, n_features + 1):\n",
    "    command = command_template.substitute(PHENOTYPE=f\"feature_{feature_idx:02}\")\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    try:\n",
    "        result.check_returncode()\n",
    "    except Exception as e:\n",
    "        print(result.stderr.decode())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "376de6ee-777a-4616-918d-b84486cf6145",
   "metadata": {},
   "source": [
    "### Projections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "520979db-633e-4da8-950c-a4de8b923b1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7eecd18da9d54afdbce0f5f308f53a1e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: optparse\n",
      "Error in glmmkin.ai_PCG_Rcpp_Quantitative(bedFile, bimFile, famFile, Xorig,  : \n",
      "  ERROR! The first variance component parameter estimate is 0\n",
      "Calls: fitNULLGLMM -> system.time -> glmmkin.ai_PCG_Rcpp_Quantitative\n",
      "Timing stopped at: 10.55 0.107 1.695\n",
      "Execution halted\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    wzhou88/saige:1.3.0 step1_fitNULLGLMM.R \\\n",
    "    --plinkFile=data/simulated-phenotypes/Genotypes \\\n",
    "    --useSparseGRMtoFitNULL=FALSE \\\n",
    "    --covarColList=sharedConfounder_norm1,sharedConfounder_norm2,sharedConfounder_norm3,independentConfounder_norm1,independentConfounder_norm2 \\\n",
    "    --sampleIDColinphenoFile=IID \\\n",
    "    --invNormalize=TRUE \\\n",
    "    --traitType=quantitative \\\n",
    "    --nThreads=16 \\\n",
    "    --IsOverwriteVarianceRatioFile=TRUE \\\n",
    "    --phenoCol=$PHENOTYPE \\\n",
    "    --phenoFile=data/projected-phenotypes/phenotypes_saige.tsv \\\n",
    "    --outputPrefix=data/gwas/saige/projections/$PHENOTYPE\n",
    "\"\"\")\n",
    "\n",
    "for projection_idx in tqdm.tnrange(1, n_projections + 1):\n",
    "    command = command_template.substitute(PHENOTYPE=f\"projection_{projection_idx:03}\")\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    try:\n",
    "        result.check_returncode()\n",
    "    except Exception as e:\n",
    "        print(result.stderr.decode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fe8692f9-0ec3-4d8b-9f55-e8835685b638",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3c6d1f572d7147eb94bc40edd5ae37cb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: RhpcBLASctl\n",
      "Error in load(GMMATmodelFile) : empty (zero-byte) input file\n",
      "Calls: SPAGMMATtest -> ReadModel -> load\n",
      "Execution halted\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "docker run \\\n",
    "    -v ./data:/data -w / \\\n",
    "    wzhou88/saige:1.3.0 step2_SPAtests.R \\\n",
    "    --bedFile=data/simulated-phenotypes/Genotypes.bed \\\n",
    "    --bimFile=data/simulated-phenotypes/Genotypes.bim \\\n",
    "    --famFile=data/simulated-phenotypes/Genotypes.fam \\\n",
    "    --LOCO=FALSE \\\n",
    "    --GMMATmodelFile=data/gwas/saige/projections/$PHENOTYPE.rda \\\n",
    "    --varianceRatioFile=data/gwas/saige/projections/$PHENOTYPE.varianceRatio.txt \\\n",
    "    --SAIGEOutputFile=data/gwas/saige/projections/$PHENOTYPE.txt\n",
    "\"\"\")\n",
    "\n",
    "for projection_idx in tqdm.tnrange(1, n_projections + 1):\n",
    "    command = command_template.substitute(PHENOTYPE=f\"projection_{projection_idx:03}\")\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    try:\n",
    "        result.check_returncode()\n",
    "    except Exception as e:\n",
    "        print(result.stderr.decode())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2730443-84f2-44e8-a3c6-2f1b671adcdc",
   "metadata": {},
   "source": [
    "## FastGWA\n",
    "\n",
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "44324eb5-1637-4b6b-8d4c-0504a5e0e6f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d4fdb73cdb1b449aa36e7121ee676a2d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "gcta64 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --grm-sparse data/grm/gcta-sparse \\\n",
    "    --pheno data/simulated-phenotypes/Ysim_plink.txt \\\n",
    "    --mpheno $mpheno \\\n",
    "    --qcovar data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --fastGWA-mlm \\\n",
    "    --thread-num 16 \\\n",
    "    --est-vg HE \\\n",
    "    --save-fastGWA-mlm-residual \\\n",
    "    --out data/gwas/fastgwa/features/result_$mpheno\n",
    "\"\"\")\n",
    "\n",
    "for feature_idx in tqdm.tnrange(1, n_features + 1):\n",
    "    command = command_template.substitute(mpheno=feature_idx)\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    result.check_returncode()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0a1ada8-b1eb-481f-bc9a-ba131df3d557",
   "metadata": {},
   "source": [
    "### Projections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9c290192-1069-492e-9e70-7de83141ca12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4e83fb1d2ae34a71a695821e0b01c5b3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "command_template = Template(\"\"\"\n",
    "gcta64 \\\n",
    "    --bfile data/simulated-phenotypes/Genotypes \\\n",
    "    --grm-sparse data/grm/gcta-sparse \\\n",
    "    --pheno data/projected-phenotypes/phenotypes.tsv \\\n",
    "    --mpheno $mpheno \\\n",
    "    --qcovar data/simulated-phenotypes/Covs_plink.txt \\\n",
    "    --fastGWA-mlm \\\n",
    "    --thread-num 16 \\\n",
    "    --est-vg HE \\\n",
    "    --save-fastGWA-mlm-residual \\\n",
    "    --out data/gwas/fastgwa/projections/result_$mpheno\n",
    "\"\"\")\n",
    "\n",
    "for feature_idx in tqdm.tnrange(1, n_projections + 1):\n",
    "    command = command_template.substitute(mpheno=feature_idx)\n",
    "    result = subprocess.run(shlex.split(command), capture_output=True)\n",
    "    result.check_returncode()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
