{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc557ef0-82d3-4cb0-9a72-69ca4a8fb5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import shlex\n",
    "\n",
    "import polars as pl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "43d6a2d7-0f85-4ccb-83fe-d3920e1c1275",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>IID</th><th>feature_01</th><th>feature_02</th><th>c1</th><th>c2</th><th>c3</th><th>c4</th><th>c5</th></tr><tr><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;ID_1&quot;</td><td>-1.946783</td><td>-1.207065</td><td>1.079511</td><td>0.675817</td><td>0.335099</td><td>1.812403</td><td>-0.383385</td></tr><tr><td>&quot;ID_2&quot;</td><td>-2.037948</td><td>0.431467</td><td>2.308693</td><td>-0.211148</td><td>-1.933398</td><td>1.139712</td><td>0.087484</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2, 8)\n",
       "┌──────┬────────────┬────────────┬──────────┬───────────┬───────────┬──────────┬───────────┐\n",
       "│ IID  ┆ feature_01 ┆ feature_02 ┆ c1       ┆ c2        ┆ c3        ┆ c4       ┆ c5        │\n",
       "│ ---  ┆ ---        ┆ ---        ┆ ---      ┆ ---       ┆ ---       ┆ ---      ┆ ---       │\n",
       "│ str  ┆ f64        ┆ f64        ┆ f64      ┆ f64       ┆ f64       ┆ f64      ┆ f64       │\n",
       "╞══════╪════════════╪════════════╪══════════╪═══════════╪═══════════╪══════════╪═══════════╡\n",
       "│ ID_1 ┆ -1.946783  ┆ -1.207065  ┆ 1.079511 ┆ 0.675817  ┆ 0.335099  ┆ 1.812403 ┆ -0.383385 │\n",
       "│ ID_2 ┆ -2.037948  ┆ 0.431467   ┆ 2.308693 ┆ -0.211148 ┆ -1.933398 ┆ 1.139712 ┆ 0.087484  │\n",
       "└──────┴────────────┴────────────┴──────────┴───────────┴───────────┴──────────┴───────────┘"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_df = (\n",
    "    pl.read_csv(\"data/simulated-phenotypes/Ysim_saige.txt\", separator=\"\\t\")\n",
    "    .select(\"IID\", \"feature_01\", \"feature_02\", pl.col(\"^.+Confounder.+$\"))\n",
    "    .rename({\n",
    "        \"sharedConfounder_norm1\": \"c1\",\n",
    "        \"sharedConfounder_norm2\": \"c2\",\n",
    "        \"sharedConfounder_norm3\": \"c3\",\n",
    "        \"independentConfounder_norm1\": \"c4\",\n",
    "        \"independentConfounder_norm2\": \"c5\"\n",
    "    })\n",
    ")\n",
    "pheno_df.write_csv(\"debug/pheno.tsv\", separator=\"\\t\")\n",
    "pheno_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "99b62307-6ba3-4f89-9a75-2bd02b8ceeff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>feature_01_mean</th><th>feature_02_mean</th><th>feature_01_var</th><th>feature_02_var</th></tr><tr><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>3.3926e-15</td><td>-1.3619e-14</td><td>1.0</td><td>1.0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 4)\n",
       "┌─────────────────┬─────────────────┬────────────────┬────────────────┐\n",
       "│ feature_01_mean ┆ feature_02_mean ┆ feature_01_var ┆ feature_02_var │\n",
       "│ ---             ┆ ---             ┆ ---            ┆ ---            │\n",
       "│ f64             ┆ f64             ┆ f64            ┆ f64            │\n",
       "╞═════════════════╪═════════════════╪════════════════╪════════════════╡\n",
       "│ 3.3926e-15      ┆ -1.3619e-14     ┆ 1.0            ┆ 1.0            │\n",
       "└─────────────────┴─────────────────┴────────────────┴────────────────┘"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    pheno_df\n",
    "    .select(\n",
    "        pl.col(\"feature_01\", \"feature_02\").mean().name.suffix(\"_mean\"),\n",
    "        pl.col(\"feature_01\", \"feature_02\").var().name.suffix(\"_var\"),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "09f69cd4-7931-465b-9c01-38b01528ad89",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: optparse\n",
      "R version 3.6.3 (2020-02-29)\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 20.04.4 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
      "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
      "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] optparse_1.7.3 SAIGE_1.3.0   \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_3.6.3     Matrix_1.6-1       Rcpp_1.0.7         getopt_1.20.3     \n",
      "[5] grid_3.6.3         data.table_1.12.8  RcppParallel_5.1.7 lattice_0.20-40   \n",
      "$plinkFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"\"\n",
      "\n",
      "$famFile\n",
      "[1] \"\"\n",
      "\n",
      "$phenoFile\n",
      "[1] \"debug/pheno.tsv\"\n",
      "\n",
      "$phenoCol\n",
      "[1] \"feature_01\"\n",
      "\n",
      "$traitType\n",
      "[1] \"quantitative\"\n",
      "\n",
      "$invNormalize\n",
      "[1] FALSE\n",
      "\n",
      "$covarColList\n",
      "[1] \"c1,c2,c3,c4,c5\"\n",
      "\n",
      "$qCovarColList\n",
      "[1] \"\"\n",
      "\n",
      "$sampleIDColinphenoFile\n",
      "[1] \"IID\"\n",
      "\n",
      "$tol\n",
      "[1] 0.02\n",
      "\n",
      "$maxiter\n",
      "[1] 20\n",
      "\n",
      "$tolPCG\n",
      "[1] 1e-05\n",
      "\n",
      "$maxiterPCG\n",
      "[1] 500\n",
      "\n",
      "$nThreads\n",
      "[1] 16\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$numRandomMarkerforVarianceRatio\n",
      "[1] 30\n",
      "\n",
      "$skipModelFitting\n",
      "[1] FALSE\n",
      "\n",
      "$skipVarianceRatioEstimation\n",
      "[1] FALSE\n",
      "\n",
      "$memoryChunk\n",
      "[1] 2\n",
      "\n",
      "$tauInit\n",
      "[1] \"0,0\"\n",
      "\n",
      "$LOCO\n",
      "[1] TRUE\n",
      "\n",
      "$isLowMemLOCO\n",
      "[1] FALSE\n",
      "\n",
      "$traceCVcutoff\n",
      "[1] 0.0025\n",
      "\n",
      "$nrun\n",
      "[1] 30\n",
      "\n",
      "$ratioCVcutoff\n",
      "[1] 0.001\n",
      "\n",
      "$outputPrefix\n",
      "[1] \"debug/feature_01\"\n",
      "\n",
      "$outputPrefix_varRatio\n",
      "[1] \"\"\n",
      "\n",
      "$IsOverwriteVarianceRatioFile\n",
      "[1] TRUE\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$isCateVarianceRatio\n",
      "[1] FALSE\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$isCovariateTransform\n",
      "[1] TRUE\n",
      "\n",
      "$isDiagofKinSetAsOne\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMtoFitNULL\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$minMAFforGRM\n",
      "[1] 0.01\n",
      "\n",
      "$maxMissingRateforGRM\n",
      "[1] 0.15\n",
      "\n",
      "$minCovariateCount\n",
      "[1] -1\n",
      "\n",
      "$includeNonautoMarkersforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$FemaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$MaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$sexCol\n",
      "[1] \"\"\n",
      "\n",
      "$FemaleCode\n",
      "[1] \"1\"\n",
      "\n",
      "$MaleCode\n",
      "[1] \"0\"\n",
      "\n",
      "$isCovariateOffset\n",
      "[1] TRUE\n",
      "\n",
      "$SampleIDIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "tauInit  is  0 0 \n",
      "cateVarRatioMinMACVecExclude  is  10 20.5 \n",
      "cateVarRatioMaxMACVecInclude  is  20.5 \n",
      "The variance ratio file  debug/feature_01.varianceRatio.txt  already exists. IsOverwriteVarianceRatioFile=TRUE so the file will be overwritten\n",
      "16  threads will be used  \n",
      "WARNING: Genetic variants needs to be ordered by chromosome and position in the Plink file\n",
      "chromosomeStartIndexVec:  0 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "chromosomeEndIndexVec:  1999 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "WARNING: The number of autosomal chromosomes is less than 2 and leave-one-chromosome-out can't be conducted! \n",
      "1000  samples have genotypes\n",
      "formula is  feature_01~c1+c2+c3+c4+c5 \n",
      "1000  samples have non-missing phenotypes\n",
      "1000  samples will be used for analysis\n",
      "     IID X.Intercept.         c1         c2         c3        c4          c5\n",
      "1   ID_1            1  1.0795113  0.6758174  0.3350990 1.8124026 -0.38338539\n",
      "113 ID_2            1  2.3086935 -0.2111481 -1.9333983 1.1397118  0.08748401\n",
      "224 ID_3            1  0.3981491 -2.1860111  0.3261955 0.4929142  0.29802132\n",
      "335 ID_4            1  0.9380872  0.3154117  2.7557429 0.5365669  0.63462912\n",
      "446 ID_5            1  0.5725300  0.7778163  0.4009758 2.0752282 -0.31290742\n",
      "557 ID_6            1 -1.5042580  0.7007885 -0.7601302 0.2256489  0.66372084\n",
      "    feature_01 IndexPheno IndexGeno\n",
      "1   -1.9467830          1         1\n",
      "113 -2.0379484          2         2\n",
      "224 -0.9937899          3         3\n",
      "335 -0.4897420          4         4\n",
      "446 -0.3971714          5         5\n",
      "557 -0.6010324          6         6\n",
      "qr transformation has been performed on covariates\n",
      "colnames(data.new) is  feature_01 minus1 c1 c2 c3 c4 c5 \n",
      "out.transform$Param.transform$qrr:  6 6 \n",
      "[1] \"isCovariateOffset=TRUE, so fixed effects coefficients won't be estimated.\"\n",
      "geno.g_minMACVarRatio 20\n",
      "geno.g_maxMACVarRatio -1\n",
      "Markers in the Plink file with MAF <  0.01  will be removed before constructing GRM\n",
      "Markers in the Plink file with missing rate >  0.15  will be removed before constructing GRM\n",
      "feature_01  is a quantitative trait\n",
      "glm:\n",
      "\n",
      "Call:  glm(formula = formula.new, family = gaussian(link = \"identity\"), \n",
      "    data = data.new, offset = covoffset)\n",
      "\n",
      "Coefficients:\n",
      "(Intercept)  \n",
      "  3.356e-15  \n",
      "\n",
      "Degrees of Freedom: 999 Total (i.e. Null);  999 Residual\n",
      "Null Deviance:\t    828.9 \n",
      "Residual Deviance: 828.9 \tAIC: 2654\n",
      "   user  system elapsed \n",
      "  1.048   0.038   0.977 \n",
      "Start fitting the NULL GLMM\n",
      "   user  system elapsed \n",
      "  1.090   0.038   1.019 \n",
      "[1] \"Start reading genotype plink file here\"\n",
      "nbyte: 250\n",
      "nbyte: 250\n",
      "reserve: 504000\n",
      "\n",
      "M: 2000, N: 1000\n",
      "setgeno mark1\n",
      "setgeno mark2\n",
      "1206 markers with MAF >= 0.01 and missing rate <= 0.15\n",
      "time: 20.848\n",
      "[1] \"Genotype reading is done\"\n",
      "[1] \"Genotype reading is done\"\n",
      "Fixed-effect coefficients:\n",
      " (Intercept) \n",
      "3.356357e-15 \n",
      "initial tau is  1 0 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 1 0\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -9.536744e-10\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Variance component estimates:\n",
      "[1] 0.8296525 0.0000000\n",
      "\n",
      "Iteration  1 :\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 0.8296525 0.0000000\n",
      "Fixed-effect coefficients:\n",
      "            [,1]\n",
      "[1,] 4.54951e-09\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "cov:  0.0008296533 \n",
      "Variance component estimates:\n",
      "[1] 0.2623500 0.3764581\n",
      "Fixed-effect coefficients:\n",
      "            [,1]\n",
      "[1,] 4.54951e-09\n",
      "\n",
      "Iteration  2 :\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 9\n",
      "Tau:\n",
      "[1] 0.2623500 0.3764581\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -1.602258e-06\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 12\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "cov:  0.00026235 \n",
      "Variance component estimates:\n",
      "[1] 0.3328739 0.4476166\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -1.602258e-06\n",
      "\n",
      "Iteration  3 :\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 9\n",
      "Tau:\n",
      "[1] 0.3328739 0.4476166\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] -1.25775e-06\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "cov:  0.0003328739 \n",
      "Variance component estimates:\n",
      "[1] 0.3566876 0.4623950\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] -1.25775e-06\n",
      "\n",
      "Iteration  4 :\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 9\n",
      "Tau:\n",
      "[1] 0.3566876 0.4623950\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -1.104682e-06\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "cov:  0.0003566876 \n",
      "Variance component estimates:\n",
      "[1] 0.3585186 0.4625778\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -1.104682e-06\n",
      "\n",
      "Final  0.3585186 0.4625778 :\n",
      "t_end_null - t_begin, fitting the NULL model without LOCO took\n",
      "   user  system elapsed \n",
      "  3.676   0.499   0.485 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 9\n",
      "Tau:\n",
      "[1] 0.3585186 0.4625778\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -1.076674e-06\n",
      "   user  system elapsed \n",
      "  4.887   0.544   1.578 \n",
      "t_end - t_begin, fitting the NULL model took\n",
      "   user  system elapsed \n",
      "  3.839   0.506   0.601 \n",
      "Start estimating variance ratios\n",
      "\n",
      "Family: gaussian \n",
      "Link function: identity \n",
      "\n",
      "iter from getPCG1ofSigmaAndVector 9\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "length(MACvector):  1206 \n",
      "length(MACvector_forVarRatio):  794 \n",
      "Only one variance ratio will be estimated using randomly selected markers with MAC >= 20\n",
      "332 th marker in geno  1 \n",
      "MAC:  389 \n",
      "G0 0 1 2 2 1 2 2 2 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "338 th marker in geno  1 \n",
      "MAC:  183 \n",
      "G0 2 2 2 1 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "792 th marker in geno  1 \n",
      "MAC:  220 \n",
      "G0 2 2 2 1 2 2 1 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "194 th marker in geno  1 \n",
      "MAC:  800 \n",
      "G0 1 1 1 0 1 2 0 0 0 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "248 th marker in geno  1 \n",
      "MAC:  193 \n",
      "G0 2 2 2 1 2 1 2 2 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "238 th marker in geno  1 \n",
      "MAC:  418 \n",
      "G0 2 1 1 0 2 2 2 1 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "717 th marker in geno  1 \n",
      "MAC:  396 \n",
      "G0 1 2 1 2 2 2 2 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "434 th marker in geno  1 \n",
      "MAC:  221 \n",
      "G0 2 1 1 2 2 2 2 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "771 th marker in geno  1 \n",
      "MAC:  402 \n",
      "G0 2 1 1 1 2 2 1 2 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "168 th marker in geno  1 \n",
      "MAC:  789 \n",
      "G0 1 0 0 0 1 2 1 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "197 th marker in geno  1 \n",
      "MAC:  400 \n",
      "G0 2 2 1 1 2 1 2 1 2 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "91 th marker in geno  1 \n",
      "MAC:  186 \n",
      "G0 2 2 2 2 2 1 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "183 th marker in geno  1 \n",
      "MAC:  388 \n",
      "G0 1 2 1 2 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "425 th marker in geno  1 \n",
      "MAC:  822 \n",
      "G0 1 1 1 1 1 0 2 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "669 th marker in geno  1 \n",
      "MAC:  769 \n",
      "G0 1 1 0 1 2 0 2 1 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "260 th marker in geno  1 \n",
      "MAC:  216 \n",
      "G0 2 2 2 2 1 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "510 th marker in geno  1 \n",
      "MAC:  194 \n",
      "G0 2 2 2 2 2 1 2 1 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "551 th marker in geno  1 \n",
      "MAC:  796 \n",
      "G0 1 2 1 1 1 2 1 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "794 th marker in geno  1 \n",
      "MAC:  818 \n",
      "G0 1 1 1 1 1 1 1 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "198 th marker in geno  1 \n",
      "MAC:  853 \n",
      "G0 1 2 2 1 2 1 2 0 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "683 th marker in geno  1 \n",
      "MAC:  185 \n",
      "G0 2 2 2 2 1 2 1 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "188 th marker in geno  1 \n",
      "MAC:  200 \n",
      "G0 2 2 2 2 2 1 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "509 th marker in geno  1 \n",
      "MAC:  786 \n",
      "G0 2 2 0 1 1 0 2 1 0 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "229 th marker in geno  1 \n",
      "MAC:  822 \n",
      "G0 1 2 1 0 1 2 1 1 2 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "28 th marker in geno  1 \n",
      "MAC:  787 \n",
      "G0 2 1 0 2 2 0 0 0 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "496 th marker in geno  1 \n",
      "MAC:  823 \n",
      "G0 2 1 2 0 2 2 1 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "494 th marker in geno  1 \n",
      "MAC:  762 \n",
      "G0 0 1 1 0 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "219 th marker in geno  1 \n",
      "MAC:  783 \n",
      "G0 2 2 0 0 1 1 0 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "603 th marker in geno  1 \n",
      "MAC:  775 \n",
      "G0 2 1 1 1 1 1 1 2 1 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 11\n",
      "49 th marker in geno  1 \n",
      "MAC:  207 \n",
      "G0 2 2 2 2 2 2 1 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 10\n",
      "CV for variance ratio estimate using  30  markers is  0.0006289039  <  0.001 \n",
      "varRatio_null 1.527937 \n",
      "        V1   V2 V3\n",
      "1 1.527937 null  1\n",
      "closed the plinkFile!\n"
     ]
    }
   ],
   "source": [
    "!docker run \\\n",
    "    -v ./data:/data \\\n",
    "    -v ./debug:/debug \\\n",
    "    -w / \\\n",
    "    wzhou88/saige:1.3.0 step1_fitNULLGLMM.R \\\n",
    "    --phenoCol=feature_01 \\\n",
    "    --outputPrefix=debug/feature_01 \\\n",
    "    --plinkFile=data/simulated-phenotypes/Genotypes \\\n",
    "    --useSparseGRMtoFitNULL=FALSE \\\n",
    "    --phenoFile=debug/pheno.tsv \\\n",
    "    --covarColList=c1,c2,c3,c4,c5 \\\n",
    "    --sampleIDColinphenoFile=IID \\\n",
    "    --invNormalize=FALSE \\\n",
    "    --traitType=quantitative \\\n",
    "    --nThreads=16 \\\n",
    "    --IsOverwriteVarianceRatioFile=TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f619e79e-d781-4f47-a5fc-48ea549021df",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: optparse\n",
      "R version 3.6.3 (2020-02-29)\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 20.04.4 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
      "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
      "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] optparse_1.7.3 SAIGE_1.3.0   \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_3.6.3     Matrix_1.6-1       Rcpp_1.0.7         getopt_1.20.3     \n",
      "[5] grid_3.6.3         data.table_1.12.8  RcppParallel_5.1.7 lattice_0.20-40   \n",
      "$plinkFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"\"\n",
      "\n",
      "$famFile\n",
      "[1] \"\"\n",
      "\n",
      "$phenoFile\n",
      "[1] \"debug/pheno.tsv\"\n",
      "\n",
      "$phenoCol\n",
      "[1] \"feature_02\"\n",
      "\n",
      "$traitType\n",
      "[1] \"quantitative\"\n",
      "\n",
      "$invNormalize\n",
      "[1] FALSE\n",
      "\n",
      "$covarColList\n",
      "[1] \"c1,c2,c3,c4,c5\"\n",
      "\n",
      "$qCovarColList\n",
      "[1] \"\"\n",
      "\n",
      "$sampleIDColinphenoFile\n",
      "[1] \"IID\"\n",
      "\n",
      "$tol\n",
      "[1] 0.02\n",
      "\n",
      "$maxiter\n",
      "[1] 20\n",
      "\n",
      "$tolPCG\n",
      "[1] 1e-05\n",
      "\n",
      "$maxiterPCG\n",
      "[1] 500\n",
      "\n",
      "$nThreads\n",
      "[1] 16\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$numRandomMarkerforVarianceRatio\n",
      "[1] 30\n",
      "\n",
      "$skipModelFitting\n",
      "[1] FALSE\n",
      "\n",
      "$skipVarianceRatioEstimation\n",
      "[1] FALSE\n",
      "\n",
      "$memoryChunk\n",
      "[1] 2\n",
      "\n",
      "$tauInit\n",
      "[1] \"0,0\"\n",
      "\n",
      "$LOCO\n",
      "[1] TRUE\n",
      "\n",
      "$isLowMemLOCO\n",
      "[1] FALSE\n",
      "\n",
      "$traceCVcutoff\n",
      "[1] 0.0025\n",
      "\n",
      "$nrun\n",
      "[1] 30\n",
      "\n",
      "$ratioCVcutoff\n",
      "[1] 0.001\n",
      "\n",
      "$outputPrefix\n",
      "[1] \"debug/feature_02\"\n",
      "\n",
      "$outputPrefix_varRatio\n",
      "[1] \"\"\n",
      "\n",
      "$IsOverwriteVarianceRatioFile\n",
      "[1] TRUE\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$isCateVarianceRatio\n",
      "[1] FALSE\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$isCovariateTransform\n",
      "[1] TRUE\n",
      "\n",
      "$isDiagofKinSetAsOne\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMtoFitNULL\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$minMAFforGRM\n",
      "[1] 0.01\n",
      "\n",
      "$maxMissingRateforGRM\n",
      "[1] 0.15\n",
      "\n",
      "$minCovariateCount\n",
      "[1] -1\n",
      "\n",
      "$includeNonautoMarkersforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$FemaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$MaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$sexCol\n",
      "[1] \"\"\n",
      "\n",
      "$FemaleCode\n",
      "[1] \"1\"\n",
      "\n",
      "$MaleCode\n",
      "[1] \"0\"\n",
      "\n",
      "$isCovariateOffset\n",
      "[1] TRUE\n",
      "\n",
      "$SampleIDIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "tauInit  is  0 0 \n",
      "cateVarRatioMinMACVecExclude  is  10 20.5 \n",
      "cateVarRatioMaxMACVecInclude  is  20.5 \n",
      "The variance ratio file  debug/feature_02.varianceRatio.txt  already exists. IsOverwriteVarianceRatioFile=TRUE so the file will be overwritten\n",
      "16  threads will be used  \n",
      "WARNING: Genetic variants needs to be ordered by chromosome and position in the Plink file\n",
      "chromosomeStartIndexVec:  0 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "chromosomeEndIndexVec:  1999 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "WARNING: The number of autosomal chromosomes is less than 2 and leave-one-chromosome-out can't be conducted! \n",
      "1000  samples have genotypes\n",
      "formula is  feature_02~c1+c2+c3+c4+c5 \n",
      "1000  samples have non-missing phenotypes\n",
      "1000  samples will be used for analysis\n",
      "     IID X.Intercept.         c1         c2         c3        c4          c5\n",
      "1   ID_1            1  1.0795113  0.6758174  0.3350990 1.8124026 -0.38338539\n",
      "113 ID_2            1  2.3086935 -0.2111481 -1.9333983 1.1397118  0.08748401\n",
      "224 ID_3            1  0.3981491 -2.1860111  0.3261955 0.4929142  0.29802132\n",
      "335 ID_4            1  0.9380872  0.3154117  2.7557429 0.5365669  0.63462912\n",
      "446 ID_5            1  0.5725300  0.7778163  0.4009758 2.0752282 -0.31290742\n",
      "557 ID_6            1 -1.5042580  0.7007885 -0.7601302 0.2256489  0.66372084\n",
      "    feature_02 IndexPheno IndexGeno\n",
      "1    -1.207065          1         1\n",
      "113   0.431467          2         2\n",
      "224  -1.052172          3         3\n",
      "335   0.787003          4         4\n",
      "446  -1.133401          5         5\n",
      "557  -1.001711          6         6\n",
      "qr transformation has been performed on covariates\n",
      "colnames(data.new) is  feature_02 minus1 c1 c2 c3 c4 c5 \n",
      "out.transform$Param.transform$qrr:  6 6 \n",
      "[1] \"isCovariateOffset=TRUE, so fixed effects coefficients won't be estimated.\"\n",
      "geno.g_minMACVarRatio 20\n",
      "geno.g_maxMACVarRatio -1\n",
      "Markers in the Plink file with MAF <  0.01  will be removed before constructing GRM\n",
      "Markers in the Plink file with missing rate >  0.15  will be removed before constructing GRM\n",
      "feature_02  is a quantitative trait\n",
      "glm:\n",
      "\n",
      "Call:  glm(formula = formula.new, family = gaussian(link = \"identity\"), \n",
      "    data = data.new, offset = covoffset)\n",
      "\n",
      "Coefficients:\n",
      "(Intercept)  \n",
      " -1.367e-14  \n",
      "\n",
      "Degrees of Freedom: 999 Total (i.e. Null);  999 Residual\n",
      "Null Deviance:\t    973.9 \n",
      "Residual Deviance: 973.9 \tAIC: 2815\n",
      "   user  system elapsed \n",
      "  1.046   0.051   0.986 \n",
      "Start fitting the NULL GLMM\n",
      "   user  system elapsed \n",
      "  1.089   0.051   1.029 \n",
      "[1] \"Start reading genotype plink file here\"\n",
      "nbyte: 250\n",
      "nbyte: 250\n",
      "reserve: 504000\n",
      "\n",
      "M: 2000, N: 1000\n",
      "setgeno mark1\n",
      "setgeno mark2\n",
      "1206 markers with MAF >= 0.01 and missing rate <= 0.15\n",
      "time: 20.817\n",
      "[1] \"Genotype reading is done\"\n",
      "[1] \"Genotype reading is done\"\n",
      "Fixed-effect coefficients:\n",
      "  (Intercept) \n",
      "-1.367119e-14 \n",
      "initial tau is  1 0 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 1 0\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 1.907349e-09\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Variance component estimates:\n",
      "[1] 0.9746685 0.0000000\n",
      "\n",
      "Iteration  1 :\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 0.9746685 0.0000000\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 5.112332e-09\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "cov:  0.000974667 \n",
      "Variance component estimates:\n",
      "[1] 0.5729918 0.3061726\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 5.112332e-09\n",
      "\n",
      "Iteration  2 :\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 6\n",
      "Tau:\n",
      "[1] 0.5729918 0.3061726\n",
      "Fixed-effect coefficients:\n",
      "           [,1]\n",
      "[1,] 7.9836e-07\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 7\n",
      "cov:  0.0005729918 \n",
      "Variance component estimates:\n",
      "[1] 0.5863863 0.3776684\n",
      "Fixed-effect coefficients:\n",
      "           [,1]\n",
      "[1,] 7.9836e-07\n",
      "\n",
      "Iteration  3 :\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 6\n",
      "Tau:\n",
      "[1] 0.5863863 0.3776684\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 1.550442e-06\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "cov:  0.0005863864 \n",
      "Variance component estimates:\n",
      "[1] 0.5827302 0.3927346\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 1.550442e-06\n",
      "\n",
      "Final  0.5827302 0.3927346 :\n",
      "t_end_null - t_begin, fitting the NULL model without LOCO took\n",
      "   user  system elapsed \n",
      "  2.011   0.205   0.279 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 6\n",
      "Tau:\n",
      "[1] 0.5827302 0.3927346\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 1.788632e-06\n",
      "   user  system elapsed \n",
      "  3.199   0.260   1.373 \n",
      "t_end - t_begin, fitting the NULL model took\n",
      "   user  system elapsed \n",
      "  2.153   0.209   0.387 \n",
      "Start estimating variance ratios\n",
      "\n",
      "Family: gaussian \n",
      "Link function: identity \n",
      "\n",
      "iter from getPCG1ofSigmaAndVector 6\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "length(MACvector):  1206 \n",
      "length(MACvector_forVarRatio):  794 \n",
      "Only one variance ratio will be estimated using randomly selected markers with MAC >= 20\n",
      "332 th marker in geno  1 \n",
      "MAC:  389 \n",
      "G0 0 1 2 2 1 2 2 2 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "338 th marker in geno  1 \n",
      "MAC:  183 \n",
      "G0 2 2 2 1 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "792 th marker in geno  1 \n",
      "MAC:  220 \n",
      "G0 2 2 2 1 2 2 1 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "194 th marker in geno  1 \n",
      "MAC:  800 \n",
      "G0 1 1 1 0 1 2 0 0 0 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "248 th marker in geno  1 \n",
      "MAC:  193 \n",
      "G0 2 2 2 1 2 1 2 2 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "238 th marker in geno  1 \n",
      "MAC:  418 \n",
      "G0 2 1 1 0 2 2 2 1 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "717 th marker in geno  1 \n",
      "MAC:  396 \n",
      "G0 1 2 1 2 2 2 2 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "434 th marker in geno  1 \n",
      "MAC:  221 \n",
      "G0 2 1 1 2 2 2 2 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "771 th marker in geno  1 \n",
      "MAC:  402 \n",
      "G0 2 1 1 1 2 2 1 2 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "168 th marker in geno  1 \n",
      "MAC:  789 \n",
      "G0 1 0 0 0 1 2 1 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "197 th marker in geno  1 \n",
      "MAC:  400 \n",
      "G0 2 2 1 1 2 1 2 1 2 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "91 th marker in geno  1 \n",
      "MAC:  186 \n",
      "G0 2 2 2 2 2 1 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "183 th marker in geno  1 \n",
      "MAC:  388 \n",
      "G0 1 2 1 2 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "425 th marker in geno  1 \n",
      "MAC:  822 \n",
      "G0 1 1 1 1 1 0 2 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "669 th marker in geno  1 \n",
      "MAC:  769 \n",
      "G0 1 1 0 1 2 0 2 1 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "260 th marker in geno  1 \n",
      "MAC:  216 \n",
      "G0 2 2 2 2 1 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "510 th marker in geno  1 \n",
      "MAC:  194 \n",
      "G0 2 2 2 2 2 1 2 1 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "551 th marker in geno  1 \n",
      "MAC:  796 \n",
      "G0 1 2 1 1 1 2 1 2 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "794 th marker in geno  1 \n",
      "MAC:  818 \n",
      "G0 1 1 1 1 1 1 1 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "198 th marker in geno  1 \n",
      "MAC:  853 \n",
      "G0 1 2 2 1 2 1 2 0 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "683 th marker in geno  1 \n",
      "MAC:  185 \n",
      "G0 2 2 2 2 1 2 1 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "188 th marker in geno  1 \n",
      "MAC:  200 \n",
      "G0 2 2 2 2 2 1 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "509 th marker in geno  1 \n",
      "MAC:  786 \n",
      "G0 2 2 0 1 1 0 2 1 0 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "229 th marker in geno  1 \n",
      "MAC:  822 \n",
      "G0 1 2 1 0 1 2 1 1 2 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "28 th marker in geno  1 \n",
      "MAC:  787 \n",
      "G0 2 1 0 2 2 0 0 0 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "496 th marker in geno  1 \n",
      "MAC:  823 \n",
      "G0 2 1 2 0 2 2 1 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "494 th marker in geno  1 \n",
      "MAC:  762 \n",
      "G0 0 1 1 0 2 2 2 2 2 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "219 th marker in geno  1 \n",
      "MAC:  783 \n",
      "G0 2 2 0 0 1 1 0 1 1 2 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "603 th marker in geno  1 \n",
      "MAC:  775 \n",
      "G0 2 1 1 1 1 1 1 2 1 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "49 th marker in geno  1 \n",
      "MAC:  207 \n",
      "G0 2 2 2 2 2 2 1 1 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 8\n",
      "CV for variance ratio estimate using  30  markers is  0.0004395841  <  0.001 \n",
      "varRatio_null 1.145289 \n",
      "        V1   V2 V3\n",
      "1 1.145289 null  1\n",
      "closed the plinkFile!\n"
     ]
    }
   ],
   "source": [
    "!docker run \\\n",
    "    -v ./data:/data \\\n",
    "    -v ./debug:/debug \\\n",
    "    -w / \\\n",
    "    wzhou88/saige:1.3.0 step1_fitNULLGLMM.R \\\n",
    "    --phenoCol=feature_02 \\\n",
    "    --outputPrefix=debug/feature_02 \\\n",
    "    --plinkFile=data/simulated-phenotypes/Genotypes \\\n",
    "    --useSparseGRMtoFitNULL=FALSE \\\n",
    "    --phenoFile=debug/pheno.tsv \\\n",
    "    --covarColList=c1,c2,c3,c4,c5 \\\n",
    "    --sampleIDColinphenoFile=IID \\\n",
    "    --invNormalize=FALSE \\\n",
    "    --traitType=quantitative \\\n",
    "    --nThreads=16 \\\n",
    "    --IsOverwriteVarianceRatioFile=TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7aaf9c1a-5481-4324-a5a3-b4da0e619698",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: RhpcBLASctl\n",
      "R version 3.6.3 (2020-02-29)\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 20.04.4 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
      "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
      "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] data.table_1.12.8   optparse_1.7.3      RhpcBLASctl_0.23-42\n",
      "[4] SAIGE_1.3.0        \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_3.6.3     Matrix_1.6-1       Rcpp_1.0.7         getopt_1.20.3     \n",
      "[5] grid_3.6.3         RcppParallel_5.1.7 lattice_0.20-40   \n",
      "$vcfFile\n",
      "[1] \"\"\n",
      "\n",
      "$vcfFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$vcfField\n",
      "[1] \"DS\"\n",
      "\n",
      "$savFile\n",
      "[1] \"\"\n",
      "\n",
      "$savFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFile\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$sampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.bed\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.bim\"\n",
      "\n",
      "$famFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.fam\"\n",
      "\n",
      "$AlleleOrder\n",
      "[1] \"ref-first\"\n",
      "\n",
      "$idstoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$rangestoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$chrom\n",
      "[1] \"\"\n",
      "\n",
      "$is_imputed_data\n",
      "[1] FALSE\n",
      "\n",
      "$minMAF\n",
      "[1] 0\n",
      "\n",
      "$minMAC\n",
      "[1] 0.5\n",
      "\n",
      "$minGroupMAC_in_BurdenTest\n",
      "[1] 5\n",
      "\n",
      "$minInfo\n",
      "[1] 0\n",
      "\n",
      "$maxMissing\n",
      "[1] 0.15\n",
      "\n",
      "$impute_method\n",
      "[1] \"best_guess\"\n",
      "\n",
      "$LOCO\n",
      "[1] FALSE\n",
      "\n",
      "$GMMATmodelFile\n",
      "[1] \"debug/feature_01.rda\"\n",
      "\n",
      "$varianceRatioFile\n",
      "[1] \"debug/feature_01.varianceRatio.txt\"\n",
      "\n",
      "$SAIGEOutputFile\n",
      "[1] \"debug/feature_01.txt\"\n",
      "\n",
      "$markers_per_chunk\n",
      "[1] 10000\n",
      "\n",
      "$groups_per_chunk\n",
      "[1] 100\n",
      "\n",
      "$is_output_moreDetails\n",
      "[1] TRUE\n",
      "\n",
      "$is_overwrite_output\n",
      "[1] TRUE\n",
      "\n",
      "$maxMAF_in_groupTest\n",
      "[1] \"0.0001,0.001,0.01\"\n",
      "\n",
      "$maxMAC_in_groupTest\n",
      "[1] \"0\"\n",
      "\n",
      "$annotation_in_groupTest\n",
      "[1] \"lof,missense;lof,missense;lof;synonymous\"\n",
      "\n",
      "$groupFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$MACCutoff_to_CollapseUltraRare\n",
      "[1] 10\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$weights.beta\n",
      "[1] \"1,25\"\n",
      "\n",
      "$r.corr\n",
      "[1] 0\n",
      "\n",
      "$markers_per_chunk_in_groupTest\n",
      "[1] 100\n",
      "\n",
      "$condition\n",
      "[1] \"\"\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$dosage_zerod_cutoff\n",
      "[1] 0.2\n",
      "\n",
      "$dosage_zerod_MAC_cutoff\n",
      "[1] 10\n",
      "\n",
      "$is_single_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_no_weight_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_output_markerList_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_Firth_beta\n",
      "[1] FALSE\n",
      "\n",
      "$pCutoffforFirth\n",
      "[1] 0.01\n",
      "\n",
      "$is_fastTest\n",
      "[1] FALSE\n",
      "\n",
      "$max_MAC_for_ER\n",
      "[1] 4\n",
      "\n",
      "$subSampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "[1] \"opt$r.corr\"\n",
      "[1] 0\n",
      "dosage_zerod_cutoff  0.2 \n",
      "Any dosages <=  0.2  for genetic variants with MAC <=  10  are set to be 0.\n",
      "single-variant association test will be performed\n",
      "[1] \"Leave-one-chromosome-out is not applied\"\n",
      "[1] \"LOCO = FASLE and leave-one-chromosome-out is not applied\"\n",
      "variance Ratio null is  1.527937 \n",
      "dosageFile type is  plink \n",
      "allele order in the plink file is  ref-first .\n",
      "Reading bim file....\n",
      "Reading fam file....\n",
      "Setting position of samples in PLINK files....\n",
      "Number of all markers to test:\t 2000 \n",
      "Number of markers in each chunk:\t 10000 \n",
      "Number of chunks for all markers:\t 1 \n",
      "(2024-11-13 20:05:44) ---- Analyzing Chunk 1/1: chrom InitialChunk ---- \n",
      "2000 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  1.164   0.050   1.105 \n",
      "[1] \"Analysis done! The results have been saved to 'debug/feature_01.txt'.\"\n"
     ]
    }
   ],
   "source": [
    "!docker run \\\n",
    "    -v ./data:/data \\\n",
    "    -v ./debug:/debug \\\n",
    "    -w / \\\n",
    "    wzhou88/saige:1.3.0 step2_SPAtests.R \\\n",
    "    --GMMATmodelFile=debug/feature_01.rda \\\n",
    "    --varianceRatioFile=debug/feature_01.varianceRatio.txt \\\n",
    "    --SAIGEOutputFile=debug/feature_01.txt \\\n",
    "    --bedFile=data/simulated-phenotypes/Genotypes.bed \\\n",
    "    --bimFile=data/simulated-phenotypes/Genotypes.bim \\\n",
    "    --famFile=data/simulated-phenotypes/Genotypes.fam \\\n",
    "    --AlleleOrder=ref-first \\\n",
    "    --is_output_moreDetails=TRUE \\\n",
    "    --LOCO=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e42af93a-5087-4bf0-9158-da2fb71a3d5c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested\n",
      "Loading required package: RhpcBLASctl\n",
      "R version 3.6.3 (2020-02-29)\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 20.04.4 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
      "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
      "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] data.table_1.12.8   optparse_1.7.3      RhpcBLASctl_0.23-42\n",
      "[4] SAIGE_1.3.0        \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_3.6.3     Matrix_1.6-1       Rcpp_1.0.7         getopt_1.20.3     \n",
      "[5] grid_3.6.3         RcppParallel_5.1.7 lattice_0.20-40   \n",
      "$vcfFile\n",
      "[1] \"\"\n",
      "\n",
      "$vcfFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$vcfField\n",
      "[1] \"DS\"\n",
      "\n",
      "$savFile\n",
      "[1] \"\"\n",
      "\n",
      "$savFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFile\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$sampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.bed\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.bim\"\n",
      "\n",
      "$famFile\n",
      "[1] \"data/simulated-phenotypes/Genotypes.fam\"\n",
      "\n",
      "$AlleleOrder\n",
      "[1] \"ref-first\"\n",
      "\n",
      "$idstoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$rangestoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$chrom\n",
      "[1] \"\"\n",
      "\n",
      "$is_imputed_data\n",
      "[1] FALSE\n",
      "\n",
      "$minMAF\n",
      "[1] 0\n",
      "\n",
      "$minMAC\n",
      "[1] 0.5\n",
      "\n",
      "$minGroupMAC_in_BurdenTest\n",
      "[1] 5\n",
      "\n",
      "$minInfo\n",
      "[1] 0\n",
      "\n",
      "$maxMissing\n",
      "[1] 0.15\n",
      "\n",
      "$impute_method\n",
      "[1] \"best_guess\"\n",
      "\n",
      "$LOCO\n",
      "[1] FALSE\n",
      "\n",
      "$GMMATmodelFile\n",
      "[1] \"debug/feature_02.rda\"\n",
      "\n",
      "$varianceRatioFile\n",
      "[1] \"debug/feature_02.varianceRatio.txt\"\n",
      "\n",
      "$SAIGEOutputFile\n",
      "[1] \"debug/feature_02.txt\"\n",
      "\n",
      "$markers_per_chunk\n",
      "[1] 10000\n",
      "\n",
      "$groups_per_chunk\n",
      "[1] 100\n",
      "\n",
      "$is_output_moreDetails\n",
      "[1] TRUE\n",
      "\n",
      "$is_overwrite_output\n",
      "[1] TRUE\n",
      "\n",
      "$maxMAF_in_groupTest\n",
      "[1] \"0.0001,0.001,0.01\"\n",
      "\n",
      "$maxMAC_in_groupTest\n",
      "[1] \"0\"\n",
      "\n",
      "$annotation_in_groupTest\n",
      "[1] \"lof,missense;lof,missense;lof;synonymous\"\n",
      "\n",
      "$groupFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$MACCutoff_to_CollapseUltraRare\n",
      "[1] 10\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$weights.beta\n",
      "[1] \"1,25\"\n",
      "\n",
      "$r.corr\n",
      "[1] 0\n",
      "\n",
      "$markers_per_chunk_in_groupTest\n",
      "[1] 100\n",
      "\n",
      "$condition\n",
      "[1] \"\"\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$dosage_zerod_cutoff\n",
      "[1] 0.2\n",
      "\n",
      "$dosage_zerod_MAC_cutoff\n",
      "[1] 10\n",
      "\n",
      "$is_single_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_no_weight_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_output_markerList_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_Firth_beta\n",
      "[1] FALSE\n",
      "\n",
      "$pCutoffforFirth\n",
      "[1] 0.01\n",
      "\n",
      "$is_fastTest\n",
      "[1] FALSE\n",
      "\n",
      "$max_MAC_for_ER\n",
      "[1] 4\n",
      "\n",
      "$subSampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "[1] \"opt$r.corr\"\n",
      "[1] 0\n",
      "dosage_zerod_cutoff  0.2 \n",
      "Any dosages <=  0.2  for genetic variants with MAC <=  10  are set to be 0.\n",
      "single-variant association test will be performed\n",
      "[1] \"Leave-one-chromosome-out is not applied\"\n",
      "[1] \"LOCO = FASLE and leave-one-chromosome-out is not applied\"\n",
      "variance Ratio null is  1.145289 \n",
      "dosageFile type is  plink \n",
      "allele order in the plink file is  ref-first .\n",
      "Reading bim file....\n",
      "Reading fam file....\n",
      "Setting position of samples in PLINK files....\n",
      "Number of all markers to test:\t 2000 \n",
      "Number of markers in each chunk:\t 10000 \n",
      "Number of chunks for all markers:\t 1 \n",
      "(2024-11-13 20:05:46) ---- Analyzing Chunk 1/1: chrom InitialChunk ---- \n",
      "2000 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  1.214   0.055   1.161 \n",
      "[1] \"Analysis done! The results have been saved to 'debug/feature_02.txt'.\"\n"
     ]
    }
   ],
   "source": [
    "!docker run \\\n",
    "    -v ./data:/data \\\n",
    "    -v ./debug:/debug \\\n",
    "    -w / \\\n",
    "    wzhou88/saige:1.3.0 step2_SPAtests.R \\\n",
    "    --GMMATmodelFile=debug/feature_02.rda \\\n",
    "    --varianceRatioFile=debug/feature_02.varianceRatio.txt \\\n",
    "    --SAIGEOutputFile=debug/feature_02.txt \\\n",
    "    --bedFile=data/simulated-phenotypes/Genotypes.bed \\\n",
    "    --bimFile=data/simulated-phenotypes/Genotypes.bim \\\n",
    "    --famFile=data/simulated-phenotypes/Genotypes.fam \\\n",
    "    --AlleleOrder=ref-first \\\n",
    "    --is_output_moreDetails=TRUE \\\n",
    "    --LOCO=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b3c1020-39f2-4493-b575-aaed16953f9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf21b72b-9950-4f82-8ad1-e70f4741f88a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ad5af7-dba0-4357-b14d-bdc826e5225d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
